<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to bootRanges • nullranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to bootRanges">
<meta property="og:description" content="nullranges">
<meta property="og:image" content="https://nullranges.github.io/nullranges/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/nullranges.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bootRanges.html">Introduction to bootRanges</a>
    </li>
    <li>
      <a href="../articles/matchRanges.html">Introduction to matchRanges</a>
    </li>
    <li>
      <a href="../articles/matching_granges.html">Matching case study I: CTCF occupancy</a>
    </li>
    <li>
      <a href="../articles/matching_ginteractions.html">Matching case study II: CTCF orientation</a>
    </li>
    <li>
      <a href="../articles/matching_pool_set.html">Creating a pool set for matchRanges</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nullranges/nullranges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to bootRanges</h1>
                        <h4 data-toc-skip class="author">Wancen Mu</h4>
            
            <h4 data-toc-skip class="date">06/23/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nullranges/nullranges/blob/HEAD/vignettes/bootRanges.Rmd" class="external-link"><code>vignettes/bootRanges.Rmd</code></a></small>
      <div class="hidden name"><code>bootRanges.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we demonstrate the block bootstrap functionality
implemented in <em>nullranges</em>. See the main nullranges vignette for
an overview of the idea of bootstrapping, or the diagram below.</p>
<p><em>nullranges</em> contains an implementation of a block bootstrap
for genomic data, as proposed by <span class="citation">Bickel et al.
(2010)</span>, such that features (ranges) are sampled from the genome
in blocks. The original block bootstrapping algorithm for genomic data
is implemented in a python software called Genome Structure Correlation,
<a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a>.</p>
<p>Our description of the <em>bootRanges</em> methods is described in
<span class="citation">Mu et al. (2023)</span>.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Minimal code for running <code><a href="../reference/bootRanges.html">bootRanges()</a></code> is shown below.
Genome segmentation <code>seg</code> and excluded regions
<code>exclude</code> are optional.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu">ExperimentHub</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># some default resources:</span></span>
<span><span class="va">seg</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7307"</span><span class="op">]</span><span class="op">]</span> <span class="co"># pre-built genome segmentation for hg38</span></span>
<span><span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH107305"</span><span class="op">]</span><span class="op">]</span> <span class="co"># Kundaje excluded regions for hg38, see below</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co"># set seed for reproducibility</span></span>
<span><span class="va">blockLength</span> <span class="op">&lt;-</span> <span class="fl">5e5</span> <span class="co"># size of blocks to bootstrap</span></span>
<span><span class="va">R</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co"># number of iterations of the bootstrap</span></span>
<span></span>
<span><span class="co"># input `ranges` require seqlengths, if missing see `GenomeInfoDb::Seqinfo`</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">ranges</span><span class="op">)</span></span>
<span><span class="co"># next, we remove non-standard chromosomes ...</span></span>
<span><span class="va">ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">ranges</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="co"># ... and mitochondrial genome, as these are too short</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">ranges</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">ranges</span><span class="op">)</span>, <span class="st">"MT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate bootstraps</span></span>
<span><span class="va">boots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">ranges</span>, blockLength<span class="op">=</span><span class="va">blockLength</span>, R<span class="op">=</span><span class="va">R</span>,</span>
<span>                    seg<span class="op">=</span><span class="va">seg</span>, exclude<span class="op">=</span><span class="va">exclude</span><span class="op">)</span></span>
<span><span class="co"># `boots` can then be used with plyranges commands, e.g. join_overlap_*</span></span></code></pre></div>
<p>The <code>boots</code> object will contain a column,
<code>iter</code> which marks the different bootstrap samples that were
generated. This allows for tidy analysis with <em>plyranges</em>,
e.g. counting the number of overlapping ranges, per bootstrap iteration.
For more examples of combining <code>bootRanges</code> with
<em>plyranges</em> operations, see the <a href="https://nullranges.github.io/tidy-ranges-tutorial/" class="external-link">tidy ranges
tutorial</a>.</p>
</div>
<div class="section level2">
<h2 id="method-overview">Method overview<a class="anchor" aria-label="anchor" href="#method-overview"></a>
</h2>
<p>Several algorithms are implemented in <code><a href="../reference/bootRanges.html">bootRanges()</a></code>,
including a segmented and unsegmented version, where in the former,
blocks are sampled with respect to a particular genome segmentation.
Overall, we recommend the segmented block bootstrap given the
heterogeneity of structure across the entire genome. If the purpose is
block bootstrapping ranges within a smaller set of sequences, such as
motifs within transcript sequence, then the unsegmented algorithm would
be sufficient.</p>
<p>In a segmented block bootstrap, the blocks are sampled and placed
within regions of a genome <em>segmentation</em>. That is, for a genome
segmented into states <span class="math inline">\(1,2, \dots,
S\)</span>, only blocks from state <em>s</em> will be used to sample the
ranges of state <em>s</em> in each bootstrap sample. The process can be
visualized in diagram panel (A) below, where a block with length <span class="math inline">\(L_b\)</span> is randomly sampled with replacement
from state “red” and the features (ranges) that overlap this block are
then copied to the first tile (which is in the “red” state). The
sampling is allowed across chromosome (as shown here), as long as the
two blocks are in the same state.</p>
<p>Note that <em>nullranges</em> provides both functions for generating
a genome segmentation from e.g. gene density, described below, as well
as a default segmentation for hg38 that can be used directly.</p>
<p>An example workflow of <code><a href="../reference/bootRanges.html">bootRanges()</a></code> used in combination
with <em>plyranges</em> <span class="citation">(Lee, Cook, and Lawrence
2019)</span> is diagrammed in panel (B) below, and can be summarized
as:</p>
<ol style="list-style-type: decimal">
<li>Compute statistics of interest between <em>GRanges</em> of feature
<span class="math inline">\(x\)</span> and <em>GRanges</em> of feature
<span class="math inline">\(y\)</span> to assess association in the
original data. This could be an enrichment (amount of overlap) or other
possible statistics making use of covariates associated with each
range</li>
<li>Generate bootstrap samples of <span class="math inline">\(y\)</span>: <code><a href="../reference/bootRanges.html">bootRanges()</a></code> with
optional arguments <code>seg</code> (segmentation) and
<code>exclude</code> (excluded regions as compiled by <span class="citation">Ogata et al. (2023)</span>) to create a
<em>BootRanges</em> object (<span class="math inline">\(y'\)</span>)</li>
<li>Compute the bootstrap distribution of test statistics between
<em>GRanges</em> of feature <span class="math inline">\(x\)</span> and
<span class="math inline">\(y'\)</span> using <em>plyranges</em>
(compute overlaps of all features in <span class="math inline">\(x\)</span> with all features in <span class="math inline">\(y'\)</span>, grouping by the bootstrap
sample)</li>
<li>Conpute a bootstrap p-value or <span class="math inline">\(z\)</span> test to test the null hypothesis that
there is no association between <span class="math inline">\(x\)</span>
and <span class="math inline">\(y\)</span> (e.g. that the bootstrap data
often has as high an enrichment as the observed data)</li>
</ol>
<p><img src="images/bootRanges.jpeg"><!-- --></p>
<p>In this vignette, we give an example of segmenting the hg38 genome by
Ensembl gene density, performing bootstrap sampling of peaks ranges, and
evaluating overlaps for observed peaks and bootstrap peaks. We also
provide other examples of statistics that can be computed with the
<code>bootRanges</code> framework, including a single cell multi-omics
example and a special case of bootstrapping features in one region of
the genome.</p>
<p><strong>Proportional blocks:</strong> A finally consideration is
whether the blocks used to generate the bootstrap samples should scale
proportionally to the segment state length, with the default setting of
<code>proportionLength=TRUE</code>. When blocks scale proportionally,
<code>blockLength</code> provides the maximal length of a block, while
the actual block length used for a segmentation state is proportional to
the fraction of genomic basepairs covered by that state. It is
theoretically motivated to have the blocks scale with the overall extent
of the segment state. However, in practice, if the genome segmentation
states are very heterogeneous in size (e.g. orders of magnitude
differences), then the blocks constructed via the proportional length
method for the smaller segmentation states can be too short to
effectively capture inter-range distances. We therefore recommend
proportional length blocks unless some segmentation states have a much
smaller extent than others, in which case fixed length blocks can be
used. This option is visualized on toy data at the end of this
vignette.</p>
</div>
<div class="section level2">
<h2 id="segmented-block-bootstrap">Segmented block bootstrap<a class="anchor" aria-label="anchor" href="#segmented-block-bootstrap"></a>
</h2>
<div class="section level3">
<h3 id="case-study-i-dhs">Case study I: DHS<a class="anchor" aria-label="anchor" href="#case-study-i-dhs"></a>
</h3>
<div class="section level4">
<h4 id="import-excluded-regions">Import excluded regions<a class="anchor" aria-label="anchor" href="#import-excluded-regions"></a>
</h4>
<p>To avoid placing bootstrap features into regions of the genome that
don’t typically have features, we import excluded regions including
ENCODE-produced excludable regions<span class="citation">(Amemiya,
Kundaje, and Boyle 2019)</span>, telomeres from UCSC, centromeres.
These, and other excludable sets, are assembled in the <a href="https://dozmorovlab.github.io/excluderanges/" class="external-link">excluderanges</a>
package <span class="citation">(Ogata et al. 2023)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># hg38.Kundaje.GRCh38_unified_Excludable</span></span>
<span><span class="va">exclude_1</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH107305"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># hg38.UCSC.centromere</span></span>
<span><span class="va">exclude_2</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH107354"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># hg38.UCSC.telomere</span></span>
<span><span class="va">exclude_3</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH107355"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># hg38.UCSC.short_arm</span></span>
<span><span class="va">exclude_4</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH107356"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># combine them</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">exclude</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/Views-class.html" class="external-link">trim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">exclude_1</span>, <span class="va">exclude_2</span>, <span class="va">exclude_3</span>, <span class="va">exclude_4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">exclude</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="segmentations-choices">Segmentations choices<a class="anchor" aria-label="anchor" href="#segmentations-choices"></a>
</h4>
<p>For most genomic datasets we examine, the density of ranges of
interest (e.g. ChIP- or ATAC-seq peaks) is often correlated to other
large-scale patterns of other genomic features, such as density of
genes. <span class="citation">Bickel et al. (2010)</span> therefore
proposed the idea of bootstrapping with respect to a <strong>segmented
genome</strong> given known, large-scale genomic structures such as
isochores (“larger than 300kb”).</p>
<p>A genomic segmentation can be considered if it defines large (e.g. on
the order of ∼1 Mb), relatively homogeneous segments with respect to
feature density, and the variance of the distribution of the test
statistics become stable as block length increases (see <span class="citation">Mu et al. (2023)</span> Fig 2A).</p>
<p>There are two options for choosing a segmentation, either:</p>
<ul>
<li>Use an exiting segmentation (e.g. ChromHMM, etc.) downloaded from
AnnotationHub or external to Bioconductor (BED files imported with
<em>rtracklayer</em>)</li>
<li>Perform a <em>de novo</em> segmentation of the genome using feature
density, e.g. gene density</li>
</ul>
<p><strong>Pre-built segmentations</strong></p>
<p>Given that these genome segmentation evaluations take time and
involve consideration of multiple criteria, we have provided our
recommended segmentation for hg38. <em>nullranges</em> has generated
pre-built segmentations for easy use, which were generated using code
outlined below in the <em>Segmentation by gene density</em> section.</p>
<p>Pre-built segmentations using either <em>CBS</em> or <em>HMM</em>
methods with <span class="math inline">\(L_s=2e6\)</span> considering
excludable regions can be downloaded directly from
<em>ExperimentHub</em>. We find that the segmentation and block length
(500kb) shown in the case study below could be used for most analyses of
hg38.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">seg_cbs</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7307"</span><span class="op">]</span><span class="op">]</span> <span class="co"># prefer CBS for hg38</span></span>
<span><span class="va">seg_hmm</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7308"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">seg</span> <span class="op">&lt;-</span> <span class="va">seg_cbs</span></span></code></pre></div>
<p><strong>Segmentation by gene density</strong></p>
<p>This section describes how we generated the pre-built segmentations,
such that users with a different genome can generate a segmentation for
their own purposes. First we obtain the Ensembl genes <span class="citation">(Howe et al. 2020)</span> for segmenting by gene
density. We obtain these using the <em>ensembldb</em> package <span class="citation">(Rainer, Gatto, and Weichenberger 2019)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb" class="external-link">ensembldb</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilterList.html" class="external-link">AnnotationFilterList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneIdFilter</a></span><span class="op">(</span><span class="st">"ENSG"</span>, <span class="st">"startsWith"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="va">filt</span><span class="op">)</span></span></code></pre></div>
<p>We perform some processing to align the sequences (chromosomes) of
<code>g</code> with our excluded regions and our features of interest
(DNase hypersensitive sites, or DHS, defined below).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="co"># MT is too small for bootstrapping, so must be removed</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, <span class="st">"MT"</span><span class="op">)</span></span>
<span><span class="co"># normally we would assign a new style, but for recent host issues</span></span>
<span><span class="co"># that produced vignette build problems, we use `paste0`</span></span>
<span><span class="co">## seqlevelsStyle(g) &lt;- "UCSC" </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg38"</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">sortSeqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span></span>
<span><span class="co">##  5194  3971  3010  2505  2868  2863  2867  2353  2242  2204  3235  2940  1304 </span></span>
<span><span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span></span>
<span><span class="co">##  2224  2152  2511  2995  1170  2926  1386   835  1318  2359   523</span></span></code></pre>
<p><strong>CBS segmentation</strong></p>
<p>We first demonstrate the use of a CBS segmentation as implemented in
<em>DNAcopy</em> <span class="citation">(Olshen et al. 2004)</span>.</p>
<p>We load the <em>nullranges</em> and <em>plyranges</em> packages, and
<em>patchwork</em> in order to produce grids of plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges">nullranges</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<p>We subset the excluded ranges to those which are 500 bp or larger.
The motivation for this step is to avoid segmenting the genome into many
small pieces due to an abundance of short excluded regions. Note that we
re-save the excluded ranges to <code>exclude</code>.</p>
<p>Here, and below, we need to specify <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">plyranges::filter</a></code> as
it conflicts with <code>filter</code> exported by
<em>ensembldb</em>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">exclude</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">L_s</span> <span class="op">&lt;-</span> <span class="fl">1e6</span></span>
<span><span class="va">seg_cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,</span>
<span>                          exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"cbs"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Analyzing: Sample.1</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_cbs</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/seg-cbs-1.png"><!-- --></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/seg-cbs-2.png"><!-- --></p>
<p>Note here, the default <em>ranges</em> plot shows the whole genome.
Some of the state transitions within small regions cannot be visualized.
One can look into specific regions to observe segmentation states, by
specifying the <code>region</code> argument.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr16"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">3e7</span>,<span class="fl">4e7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_cbs</span>, <span class="va">exclude</span>, type<span class="op">=</span><span class="st">"ranges"</span>, region<span class="op">=</span><span class="va">region</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/unnamed-chunk-7-1.png"><!-- --></p>
<p><strong>Alternatively: HMM segmentation</strong></p>
<p>Here we use an alternative segmentation implemented in the
<em>RcppHMM</em> CRAN package, using the <code>initGHMM</code>,
<code>learnEM</code>, and <code>viterbi</code> functions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seg_hmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,</span>
<span>                          exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"hmm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Finished at Iteration: 127 with Error: 9.70053e-06</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_hmm</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/seg-hmm-1.png"><!-- --></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/seg-hmm-2.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="running-bootranges">Running bootRanges<a class="anchor" aria-label="anchor" href="#running-bootranges"></a>
</h4>
<p>We use a set of DNase hypersensitivity sites (DHS) from the ENCODE
project <span class="citation">(ENCODE 2012)</span> in A549 cell line
(ENCSR614GWM). Here, for speed, we work with a pre-processed data object
from ExperimentHub, created using the following steps:</p>
<ul>
<li>Download ENCODE DNase hypersensitive peaks in A549 from
<em>AnnotationHub</em>
</li>
<li>Subset to standard chromosomes and remove mitochondrial DNA</li>
<li>Use a chain file from UCSC to lift ranges from hg19 to hg38</li>
<li>Sort the DHS features to be bootstrapped</li>
</ul>
<p>These steps are included in <em>nullrangesData</em> in the
<code>inst/scripts/make-dhs-data.R</code> script.</p>
<p>For speed of the vignette, we restrict to a smaller number of DHS,
filtering by the signal value. We also remove unrelated metadata columns
that we don’t need for the bootstrap analysis. Because we are interested
in signal value for DHS peaks later, we only keep this column. Consider,
when creating bootstrapped data, that you will be creating an object
<em>many times larger</em> than your original data (i.e. multipled by
<code>R</code> the number of bootstrap iterations), so
<strong>filtering</strong> down to key ranges and
<strong>selecting</strong> only the relevant metadata can help make the
analysis much more efficient.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/DHSA549Hg38.html" class="external-link">DHSA549Hg38</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signalValue</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">signalValue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6214</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span></span>
<span><span class="co">##  1436   252   108    30   148    51   184   146   155   443   436   526    20 </span></span>
<span><span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span></span>
<span><span class="co">##   197   265   214   715    20   649   142    31    19    17    10</span></span></code></pre>
<p>Now we apply a segmented block bootstrap with blocks of size 500kb,
to the peaks. Here we show generation of 50 iterations of a block
bootstrap followed by a typical overlap analysis using
<em>plyranges</em> <span class="citation">(Lee, Cook, and Lawrence
2019)</span>.</p>
<p>Note that we have already removed non-standard chromosomes and
mitochondrial chromosome, as these are typically shorter than our
desired <code>blockLength</code> (see e.g. code in Quick Start
above).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co"># for reproducibility</span></span>
<span><span class="va">R</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">blockLength</span> <span class="op">&lt;-</span> <span class="fl">5e5</span></span>
<span><span class="va">boots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, R <span class="op">=</span> <span class="va">R</span>, seg <span class="op">=</span> <span class="va">seg</span>, exclude<span class="op">=</span><span class="va">exclude</span><span class="op">)</span></span>
<span><span class="va">boots</span></span></code></pre></div>
<pre><code><span><span class="co">## BootRanges object with 310434 ranges and 3 metadata columns:</span></span>
<span><span class="co">##            seqnames            ranges strand |        id signalValue  iter</span></span>
<span><span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]     chr1     242791-242940      * |       347         120     1</span></span>
<span><span class="co">##        [2]     chr1     256031-256180      * |       348         194     1</span></span>
<span><span class="co">##        [3]     chr1     391535-391684      * |      5301         109     1</span></span>
<span><span class="co">##        [4]     chr1     421046-421195      * |      5302         106     1</span></span>
<span><span class="co">##        [5]     chr1     438186-438335      * |      5303         232     1</span></span>
<span><span class="co">##        ...      ...               ...    ... .       ...         ...   ...</span></span>
<span><span class="co">##   [310430]     chrY 27090908-27091057      * |      2133         105    50</span></span>
<span><span class="co">##   [310431]     chrY 27194968-27195117      * |      2134         128    50</span></span>
<span><span class="co">##   [310432]     chrY 27224188-27224337      * |      2135         153    50</span></span>
<span><span class="co">##   [310433]     chrY 27234153-27234302      * |      2136         125    50</span></span>
<span><span class="co">##   [310434]     chrY 27789879-27790028      * |      2116         118    50</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 24 sequences from hg38 genome</span></span></code></pre>
<p>What is returned here? The <code>bootRanges</code> function returns a
<em>BootRanges</em> object, which is a simple sub-class of
<em>GRanges</em>. The iteration (<code>iter</code>) and (optionally) the
block length (<code>blockLength</code>) are recorded as metadata
columns, accessible via <code>mcols</code>. We return the bootstrapped
ranges as <em>GRanges</em> rather than <em>GRangesList</em>, as the
former is more compatible with downstream overlap joins with
<em>plyranges</em>, where the iteration column can be used with
<code>group_by</code> to provide per bootstrap summary statistics, as
shown below.</p>
<p>Note that we use the <code>exclude</code> object from the previous
step, which does not contain small ranges. If one wanted to also avoid
generation of bootstrapped features that overlap small excluded ranges,
then omit this filtering step (use the original, complete
<code>exclude</code> feature set).</p>
</div>
<div class="section level4">
<h4 id="assessing-quality-of-bootstrap-samples">Assessing quality of bootstrap samples<a class="anchor" aria-label="anchor" href="#assessing-quality-of-bootstrap-samples"></a>
</h4>
<p>We can examine properties of permuted y over iterations, and compare
to the original y. To do so, we first add the original features as
iter=0. Then compute summaries:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>iter<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-bind.html" class="external-link">bind_ranges</a></span><span class="op">(</span><span class="va">boots</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span></span>
<span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="va">combined</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">##   iter      n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 0      <span style="text-decoration: underline;">6</span>214</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 1      <span style="text-decoration: underline;">6</span>144</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 2      <span style="text-decoration: underline;">6</span>265</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> 3      <span style="text-decoration: underline;">6</span>233</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> 4      <span style="text-decoration: underline;">6</span>291</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> 5      <span style="text-decoration: underline;">6</span>357</span></span></code></pre>
<p>We can also look at distributions of various aspects, e.g. here the
inter-feature distance of features, across a few of the bootstraps and
the original feature set y.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/ggridges/" class="external-link">ggridges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">interdist</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span></span>
<span>    <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span>,<span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">seqnames</span> <span class="op">==</span> <span class="va">y</span><span class="op">$</span><span class="va">seqnames</span>,</span>
<span>           <span class="va">x</span><span class="op">$</span><span class="va">start</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">width</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span></span>
<span>           <span class="va">y</span><span class="op">$</span><span class="va">start</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">width</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># just looking at first 3 iterations...</span></span>
<span><span class="va">combined</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">iter</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fl">0</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>iter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">!</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>interdist <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>, <span class="va">interdist</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">iter</span>, <span class="va">interdist</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">interdist</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">iter</span> <span class="op">==</span> <span class="fl">0</span>, <span class="st">"original"</span>, <span class="st">"boot"</span><span class="op">)</span>,</span>
<span>         interdist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">interdist</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">interdist</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">interdist</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, <span class="va">iter</span>, fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://wilkelab.org/ggridges/reference/geom_density_ridges.html" class="external-link">geom_density_ridges</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">stats</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">1.5</span>, y<span class="op">=</span><span class="va">iter</span>, label<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"n="</span>,<span class="va">n</span><span class="op">)</span>, fill<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span>,</span>
<span>            vjust<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/unnamed-chunk-11-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="derive-statistics-of-interest">Derive statistics of interest<a class="anchor" aria-label="anchor" href="#derive-statistics-of-interest"></a>
</h4>
<p>Suppose we have a set of features <code>x</code> and we are
interested in evaluating the enrichment of this set with the DHS. We can
calculate for example the sum observed number of overlaps for features
in <code>x</code> with DHS in whole genome (or something more
complicated, e.g. the maximum log fold change or signal value for DHS
peaks within a <code>maxgap</code> window of <code>x</code>).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fl">50</span><span class="op">:</span><span class="fl">99</span> <span class="op">*</span> <span class="fl">1e6</span>, width<span class="op">=</span><span class="fl">1e6</span><span class="op">)</span>, x_id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="statistic-i-the-total-number-of-overlaps">Statistic I: the total number of overlaps<a class="anchor" aria-label="anchor" href="#statistic-i-the-total-number-of-overlaps"></a>
</h4>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html" class="external-link">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">dhs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span> <span class="va">x</span><span class="op">$</span><span class="va">n_overlaps</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 64</span></span></code></pre>
<p>We can repeat this with the bootstrapped features using a
<code>group_by</code> command, a <code>summarize</code>, followed by a
second <code>group_by</code> and <code>summarize</code>. It may help to
step through these commands one by one to understand what the
intermediate output is.</p>
<p>Note that we need to use <code><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">tidyr::complete</a></code> in order to
fill in combinations of <code>x</code> and <code>iter</code> where the
overlap was 0.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boot_stats</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">boots</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">complete</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>sumOverlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n_overlaps</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The above code, first grouping by <code>x_id</code> and
<code>iter</code>, then subsequently by <code>iter</code> is general and
allows for more complex analysis than just mean overlap (e.g. how many
times an <code>x</code> range has 1 or more overlap, what is the mean or
max signal value for peaks overlapping ranges in <code>x</code>,
etc.).</p>
<p>If one is interested in assessing <em>feature-wise</em> statistics
instead of <em>genome-wise</em> statistics, eg.,the mean observed number
of overlaps per feature or mean base pair overlap in <code>x</code>, one
can also group by both (<code>block</code>,<code>iter</code>). 10,000
total blocks may therefore be sufficient to derive a bootstrap
distribution, avoiding the need to generate many bootstrap genomes of
data.</p>
<p>Finally we can plot a histogram. In this case, as the <code>x</code>
features were arbitrary, our observed value falls within the
distribution of sum number of overlap bootstrapped peaks with <span class="math inline">\(x\)</span>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">boot_stats</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sumOverlaps</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth<span class="op">=</span><span class="fl">5</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">n_overlaps</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/boot-histI-1.png"><!-- --></p>
</div>
<div class="section level4">
<h4 id="statistic-ii-the-sum-of-signal-value-for-nearby-peaks">Statistic II: the sum of signal value for nearby peaks<a class="anchor" aria-label="anchor" href="#statistic-ii-the-sum-of-signal-value-for-nearby-peaks"></a>
</h4>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x_obs</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">dhs</span>,maxgap<span class="op">=</span><span class="fl">1e3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x_obs</span><span class="op">$</span><span class="va">signalValue</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9503</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boot_stats</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">boots</span>,maxgap<span class="op">=</span><span class="fl">1e3</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>Signal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">signalValue</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">complete</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Signal <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>sumSignal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Signal</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Still in this case, our observed value falls within the distribution
of bootstrapped statistics.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">boot_stats</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">sumSignal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x_obs</span><span class="op">$</span><span class="va">signalValue</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/boot-histII-1.png"><!-- --></p>
</div>
</div>
<div class="section level3">
<h3 id="case-study-ii-single-cell-multi-omics">Case study II: Single cell multi-omics<a class="anchor" aria-label="anchor" href="#case-study-ii-single-cell-multi-omics"></a>
</h3>
<p>This case study provides an example of a more complex statistic of
interest. Instead of computing simple overlap, or some function of a
single metadata column, e.g. <code>signal</code> above, a statistic can
be computed on a count matrix from a <em>SummerizedExperiment</em> or
<em>SingleCellExperiment</em>. Here we provide unevaluated code for such
an example.</p>
<p>We use a trick to perform computation on a count matrix within
plyranges: we extract count matrix data from a
<em>SingleCellExperiment</em> (e.g. pseudobulk summaries of cell type
specific expression or accessibility) and save it in a GRanges’s
metadata column as a <code><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">NumericList()</a></code> format. This allows us
to perform the block bootstrap and quickly compute statistics using
<em>plyranges</em>. The case study we consider here is to assess the
correlation of gene expression and promoter peak accessibility from a
dataset of Chromium Single Cell Multiome ATAC + Gene Expression.</p>
<p>We first load the 10X <a href="https://raw.githack.com/bioFAM/MOFA2_tutorials/master/R_tutorials/10x_scRNA_scATAC.html" class="external-link">Multiome</a>
data, as compiled by the MOFA2 tutorial. The first two pre-processed
steps are included in <em>nullrangesData</em> in the
<code>inst/scripts/make-multiome-data.R</code> script. Here, we
aggregate 14 cell types across cluster-sample groups.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## split sparse count matrix into NumericList</span></span>
<span><span class="co"># sc_rna &lt;- rna_Granges[-which(rna.sd==0)] %&gt;%</span></span>
<span><span class="co">#   mutate(counts1 = NumericList(asplit(rna.scaled, 1))) %&gt;% sort()</span></span>
<span><span class="co"># sc_promoter &lt;- promoter_Granges[-which(promoter.sd==0)] %&gt;%</span></span>
<span><span class="co">#   mutate(counts2 = NumericList(asplit(promoter.scaled, 1))) %&gt;% sort()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sc_rna"</span><span class="op">)</span></span>
<span><span class="va">sc_rna</span></span>
<span><span class="va">sc_rna</span><span class="op">$</span><span class="va">counts1</span></span>
<span><span class="va">nct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sc_rna</span><span class="op">$</span><span class="va">counts1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"There are"</span>, <span class="va">nct</span>, <span class="st">"cell types"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sc_promoter"</span><span class="op">)</span></span>
<span><span class="co">## bootstrap promoters</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">sc_promoter</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">]</span> <span class="co"># pull chrom lens from USCS</span></span>
<span><span class="va">bootranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">sc_promoter</span>, blockLength <span class="op">=</span> <span class="fl">5e5</span>, R<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>There are two options for downstream pipelines for computing
statistics after generating a <em>BootRanges</em> object. One is to use
<em>plyranges</em> in downstream analysis as in the previous sections,
while another is to create a <em>tidySummarizedExperiment</em> object
used to compute the statistics of interest.</p>
<p><strong>Plyranges</strong></p>
<p>We can compute the mean correlation of the all pairs of genes and
promoter peaks over all bootstrap iterations and compare to the observed
statistic. The histogram will show that our observed value falls far
from the distribution of bootstrapped statistics, as we expected for
this dataset.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_gr</span> <span class="op">&lt;-</span> <span class="va">sc_rna</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">sc_promoter</span>, maxgap<span class="op">=</span><span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="va">nct</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">counts1</span> <span class="op">*</span> <span class="va">counts2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>meanCor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## mean correlation distribution</span></span>
<span><span class="va">cor_boot</span> <span class="op">&lt;-</span> <span class="va">sc_rna</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">bootranges</span>, maxgap<span class="op">=</span><span class="fl">1000</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># vectorized code are 10 times faster than cor(counts1, counts2) for plyranges pipeline</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="va">nct</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">counts1</span> <span class="op">*</span> <span class="va">counts2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">rho</span>, <span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>meanCor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cor_boot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">meanCor</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">cor_gr</span><span class="op">$</span><span class="va">meanCor</span>,linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>If one is interested in assessing the significance of gene-promoter
pairs per gene, one can also use <code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">plyranges::select</a></code>
(<em>gene</em>, <em>peak</em>, <em>rho</em>) and
<code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">plyranges::group_by</a></code> <em>gene</em>.</p>
<p><strong>tidySummarizedExperiment</strong></p>
<p>In order to use <em>tidySummarizedExperiment</em> to compute the same
operation, we first must create an RangedSE for each modality, and then
we can find the overlaps..</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make an SE where each row is an overlap</span></span>
<span><span class="va">makeOverlapSE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">se_rna</span>, <span class="va">se_promoter</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se_rna</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se_promoter</span><span class="op">)</span>,maxgap <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span>  <span class="va">assay_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_rna</span>, <span class="st">"rna"</span><span class="op">)</span><span class="op">[</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span>, <span class="op">]</span></span>
<span>  <span class="va">assay_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_promoter</span>, <span class="st">"promoter"</span><span class="op">)</span><span class="op">[</span> <span class="va">idx</span><span class="op">$</span><span class="va">peak</span>, <span class="op">]</span></span>
<span>  <span class="co"># this is needed to build the SE</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">assay_x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">assay_y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span> <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span> <span class="va">idx</span><span class="op">$</span><span class="va">gene</span> <span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>    assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">assay_x</span>, y<span class="op">=</span><span class="va">assay_y</span><span class="op">)</span>,</span>
<span>    rowRanges<span class="op">=</span><span class="va">idx</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># create SE for observed data</span></span>
<span><span class="va">se_rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>rna<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">sc_rna</span><span class="op">$</span><span class="va">counts1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowRanges<span class="op">=</span><span class="va">sc_rna</span><span class="op">)</span></span>
<span><span class="va">se_promoter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>promoter<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">sc_promoter</span><span class="op">$</span><span class="va">counts2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowRanges<span class="op">=</span><span class="va">sc_promoter</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">makeOverlapSE</span><span class="op">(</span><span class="va">se_rna</span>, <span class="va">se_promoter</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">.feature</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"mean correlation is"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Repeating for the bootstrapping case</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create SE for bootranges data</span></span>
<span><span class="va">se_promoter_boots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>promoter<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">bootranges</span><span class="op">$</span><span class="va">counts2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  rowRanges<span class="op">=</span><span class="va">bootranges</span><span class="op">)</span></span>
<span><span class="va">se_boots</span> <span class="op">&lt;-</span> <span class="fu">makeOverlapSE</span><span class="op">(</span><span class="va">se_rna</span>, <span class="va">se_promoter_boots</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_boots</span> <span class="op">&lt;-</span> <span class="va">se_boots</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">.feature</span>,<span class="va">iter</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># vectorized code similar to cor(counts1, counts2) for tidySE pipeline</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>                   <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="va">data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>rho <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cor_boot</span> <span class="op">&lt;-</span> <span class="va">se_boots</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>meanCor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">rho</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">cor_boot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">meanCor</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">rho</span><span class="op">)</span>,linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For more examples of combining <code>bootRanges</code> from
<em>nullranges</em> with <em>plyranges</em> piped operations, see the
relevant chapter in the <a href="https://nullranges.github.io/tidy-ranges-tutorial/bootstrap-overlap.html" class="external-link">tidy-ranges-tutorial</a>
book.</p>
</div>
<div class="section level3">
<h3 id="case-study-iii-block-bootstrap-one-region">Case study III: Block bootstrap one region<a class="anchor" aria-label="anchor" href="#case-study-iii-block-bootstrap-one-region"></a>
</h3>
<p>Generally, it makes sense to block bootstrap the entire genome at
once. This is motivated by the “tidy analysis” paradigm where loops are
avoided by stacking data into a longer format. This makes computation
more efficient in our case (as a single overlap call can be made with
all regions of interest at once, across multiple bootstrap iterations),
and it also can simplify code and avoid repetition.</p>
<p>However, in some cases, there is a single region of interest, and it
is desired to generate bootstrap data within this one region. For this,
we have a convenience function that enables bootstrap computation.</p>
<p>Suppose we have data in the following region of chromosome 1:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/DHSA549Hg38.html" class="external-link">DHSA549Hg38</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">10e6</span> <span class="op">+</span> <span class="fl">1</span>, width<span class="op">=</span><span class="fl">1e6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">10e6</span> <span class="op">+</span> <span class="fl">0</span><span class="op">:</span><span class="fl">9</span> <span class="op">*</span> <span class="fl">1e5</span> <span class="op">+</span> <span class="fl">1</span>, width<span class="op">=</span><span class="fl">1e4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-filter-overlaps.html" class="external-link">filter_by_overlaps</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>num_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html" class="external-link">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand | num_overlaps</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1 10000001-10010000      * |            0</span></span>
<span><span class="co">##    [2]     chr1 10100001-10110000      * |            2</span></span>
<span><span class="co">##    [3]     chr1 10200001-10210000      * |            1</span></span>
<span><span class="co">##    [4]     chr1 10300001-10310000      * |            0</span></span>
<span><span class="co">##    [5]     chr1 10400001-10410000      * |            5</span></span>
<span><span class="co">##    [6]     chr1 10500001-10510000      * |            2</span></span>
<span><span class="co">##    [7]     chr1 10600001-10610000      * |            0</span></span>
<span><span class="co">##    [8]     chr1 10700001-10710000      * |            0</span></span>
<span><span class="co">##    [9]     chr1 10800001-10810000      * |            1</span></span>
<span><span class="co">##   [10]     chr1 10900001-10910000      * |            1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>We can easily bootstrap data just in this region using the following
code:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/oneRegionSegment.html">oneRegionSegment</a></span><span class="op">(</span><span class="va">region</span>, seqlength<span class="op">=</span><span class="fl">248956422</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepSeqlevels</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"chr1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">y</span>, blockLength<span class="op">=</span><span class="fl">1e5</span>, R<span class="op">=</span><span class="fl">1</span>, seg<span class="op">=</span><span class="va">seg</span>,</span>
<span>                   proportionLength<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">boot</span></span></code></pre></div>
<pre><code><span><span class="co">## BootRanges object with 83 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand |  iter</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1 10001821-10001970      * |     1</span></span>
<span><span class="co">##    [2]     chr1 10002381-10002530      * |     1</span></span>
<span><span class="co">##    [3]     chr1 10003041-10003190      * |     1</span></span>
<span><span class="co">##    [4]     chr1 10003536-10003685      * |     1</span></span>
<span><span class="co">##    [5]     chr1 10004001-10004150      * |     1</span></span>
<span><span class="co">##    ...      ...               ...    ... .   ...</span></span>
<span><span class="co">##   [79]     chr1 10918882-10919031      * |     1</span></span>
<span><span class="co">##   [80]     chr1 10953082-10953231      * |     1</span></span>
<span><span class="co">##   [81]     chr1 10961702-10961851      * |     1</span></span>
<span><span class="co">##   [82]     chr1 10993742-10993891      * |     1</span></span>
<span><span class="co">##   [83]     chr1 10995862-10996011      * |     1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from hg38 genome</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>num_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html" class="external-link">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">boot</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand | num_overlaps</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1 10000001-10010000      * |            5</span></span>
<span><span class="co">##    [2]     chr1 10100001-10110000      * |            0</span></span>
<span><span class="co">##    [3]     chr1 10200001-10210000      * |            1</span></span>
<span><span class="co">##    [4]     chr1 10300001-10310000      * |            5</span></span>
<span><span class="co">##    [5]     chr1 10400001-10410000      * |            1</span></span>
<span><span class="co">##    [6]     chr1 10500001-10510000      * |            0</span></span>
<span><span class="co">##    [7]     chr1 10600001-10610000      * |            5</span></span>
<span><span class="co">##    [8]     chr1 10700001-10710000      * |            0</span></span>
<span><span class="co">##    [9]     chr1 10800001-10810000      * |            5</span></span>
<span><span class="co">##   [10]     chr1 10900001-10910000      * |            0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Here it is important to use <code>proportionLength=FALSE</code> so
that the blocks will be of the size specified and not smaller (they
would otherwise be scaled down proportional to the fraction of
<code>region</code> compared to the chromosome).</p>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-bootstrap-types">Visualizing bootstrap types<a class="anchor" aria-label="anchor" href="#visualizing-bootstrap-types"></a>
</h2>
<p>Below we present a toy example for visualizing the segmented block
bootstrap. First, we define a helper function for plotting
<em>GRanges</em> using <em>plotgardener</em> <span class="citation">(Kramer et al. 2022)</span>. A key aspect here is that
we color the original and bootstrapped ranges by the genomic state (the
state of the segmentation that the original ranges fall in).</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener" class="external-link">plotgardener</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">my_palette</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"green3"</span>,<span class="st">"red3"</span>,<span class="st">"dodgerblue"</span>,</span>
<span>         <span class="st">"blue2"</span>,<span class="st">"green4"</span>,<span class="st">"darkred"</span><span class="op">)</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">plotGRanges</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html" class="external-link">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">5</span>, xgrid <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                ygrid <span class="op">=</span> <span class="fl">0</span>, showGuides <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>    <span class="va">chromend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[[</span><span class="va">chrom</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>      <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pgParams.html" class="external-link">pgParams</a></span><span class="op">(</span>chromstart <span class="op">=</span> <span class="fl">0</span>, chromend <span class="op">=</span> <span class="va">chromend</span>,</span>
<span>                    x <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="va">chromend</span><span class="op">/</span><span class="fl">500</span>, height <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">chromend</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>                    fill <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/colorby.html" class="external-link">colorby</a></span><span class="op">(</span><span class="st">"state_col"</span>, palette<span class="op">=</span><span class="va">my_palette</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="va">prngs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotRanges.html" class="external-link">plotRanges</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gr</span>, params <span class="op">=</span> <span class="va">p</span>,</span>
<span>                          chrom <span class="op">=</span> <span class="va">chrom</span>,</span>
<span>                          y <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span>,</span>
<span>                          just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoGenomeLabel.html" class="external-link">annoGenomeLabel</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">prngs</span>, params <span class="op">=</span> <span class="va">p</span>, y <span class="op">=</span> <span class="fl">0.1</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Create a toy genome segmentation:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="va">seq_nms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="va">seq_nms</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">101</span>, <span class="fl">201</span>, <span class="fl">401</span>, <span class="fl">1</span>, <span class="fl">201</span>, <span class="fl">301</span><span class="op">)</span>,</span>
<span>          width <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  seqlengths<span class="op">=</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span>chr1<span class="op">=</span><span class="fl">500</span>,chr2<span class="op">=</span><span class="fl">400</span><span class="op">)</span>,</span>
<span>  state <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>  state_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can visualize with our helper function:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">seg</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/toysegments-1.png"><!-- --></p>
<p>Now create small ranges distributed uniformly across the toy
genome:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="va">n</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">500</span><span class="op">-</span><span class="fl">10</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr2"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">400</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">seg</span>, type<span class="op">=</span><span class="st">"within"</span>, select<span class="op">=</span><span class="st">"first"</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">idx</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">state</span> <span class="op">&lt;-</span> <span class="va">seg</span><span class="op">$</span><span class="va">state</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">state_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seg</span><span class="op">$</span><span class="va">state_col</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/toyrangesI-1.png"><!-- --></p>
<p><strong>Scaling vs. not scaling by segment length</strong></p>
<p>We can visualize block bootstrapped ranges when the blocks do not
scale to segment state length:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">25</span>, seg <span class="op">=</span> <span class="va">seg</span>,</span>
<span>                       proportionLength <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/toy-no-prop-1.png"><!-- --></p>
<p>This time the blocks scale to the segment state length. Note that in
this case <code>blockLength</code> is the <em>maximal</em> block length
possible, but the actual block lengths per segment will be smaller
(proportional to the fraction of basepairs covered by that state in the
genome segmentation).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">50</span>, seg <span class="op">=</span> <span class="va">seg</span>,</span>
<span>                       proportionLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></span></code></pre></div>
<p><img src="bootRanges_files/figure-html/toy-prop-1.png"><!-- --></p>
<p>Note that some ranges from adjacent states are allowed to be placed
within different states in the bootstrap sample. This is because, during
the random sampling of blocks of original data, a block is allowed to
extend beyond the segmentation region of the state being sampled, and
features from adjacent states are not excluded from the sampled
block.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] plotgardener_1.6.0          ggplot2_3.4.2              </span></span>
<span><span class="co">##  [3] purrr_1.0.1                 ggridges_0.5.4             </span></span>
<span><span class="co">##  [5] tidyr_1.3.0                 patchwork_1.1.2            </span></span>
<span><span class="co">##  [7] plyranges_1.20.0            nullranges_1.7.6           </span></span>
<span><span class="co">##  [9] EnsDb.Hsapiens.v86_2.99.0   ensembldb_2.24.0           </span></span>
<span><span class="co">## [11] AnnotationFilter_1.24.0     GenomicFeatures_1.52.0     </span></span>
<span><span class="co">## [13] AnnotationDbi_1.62.1        nullrangesData_1.6.0       </span></span>
<span><span class="co">## [15] InteractionSet_1.28.1       SummarizedExperiment_1.30.2</span></span>
<span><span class="co">## [17] Biobase_2.60.0              MatrixGenerics_1.12.2      </span></span>
<span><span class="co">## [19] matrixStats_1.0.0           ExperimentHub_2.8.0        </span></span>
<span><span class="co">## [21] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [23] IRanges_2.34.0              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [25] AnnotationHub_3.8.0         BiocFileCache_2.8.0        </span></span>
<span><span class="co">## [27] dbplyr_2.3.2                BiocGenerics_0.46.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] strawr_0.0.91                 RColorBrewer_1.1-3           </span></span>
<span><span class="co">##   [3] jsonlite_1.8.5                magrittr_2.0.3               </span></span>
<span><span class="co">##   [5] farver_2.1.1                  rmarkdown_2.22               </span></span>
<span><span class="co">##   [7] fs_1.6.2                      BiocIO_1.10.0                </span></span>
<span><span class="co">##   [9] zlibbioc_1.46.0               ragg_1.2.5                   </span></span>
<span><span class="co">##  [11] vctrs_0.6.3                   memoise_2.0.1                </span></span>
<span><span class="co">##  [13] Rsamtools_2.16.0              RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [15] htmltools_0.5.5               S4Arrays_1.0.4               </span></span>
<span><span class="co">##  [17] progress_1.2.2                curl_5.0.1                   </span></span>
<span><span class="co">##  [19] gridGraphics_0.5-1            sass_0.4.6                   </span></span>
<span><span class="co">##  [21] bslib_0.5.0                   desc_1.4.2                   </span></span>
<span><span class="co">##  [23] cachem_1.0.8                  GenomicAlignments_1.36.0     </span></span>
<span><span class="co">##  [25] mime_0.12                     lifecycle_1.0.3              </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3               Matrix_1.5-4.1               </span></span>
<span><span class="co">##  [29] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.10       shiny_1.7.4                  </span></span>
<span><span class="co">##  [33] digest_0.6.31                 colorspace_2.1-0             </span></span>
<span><span class="co">##  [35] rprojroot_2.0.3               textshaping_0.3.6            </span></span>
<span><span class="co">##  [37] RSQLite_2.3.1                 filelock_1.0.2               </span></span>
<span><span class="co">##  [39] labeling_0.4.2                fansi_1.0.4                  </span></span>
<span><span class="co">##  [41] httr_1.4.6                    compiler_4.3.0               </span></span>
<span><span class="co">##  [43] bit64_4.0.5                   withr_2.5.0                  </span></span>
<span><span class="co">##  [45] BiocParallel_1.34.2           DBI_1.1.3                    </span></span>
<span><span class="co">##  [47] highr_0.10                    biomaRt_2.56.1               </span></span>
<span><span class="co">##  [49] rappdirs_0.3.3                DelayedArray_0.26.3          </span></span>
<span><span class="co">##  [51] rjson_0.2.21                  DNAcopy_1.74.1               </span></span>
<span><span class="co">##  [53] tools_4.3.0                   interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [55] httpuv_1.6.11                 glue_1.6.2                   </span></span>
<span><span class="co">##  [57] restfulr_0.0.15               promises_1.2.0.1             </span></span>
<span><span class="co">##  [59] grid_4.3.0                    generics_0.1.3               </span></span>
<span><span class="co">##  [61] gtable_0.3.3                  data.table_1.14.8            </span></span>
<span><span class="co">##  [63] hms_1.1.3                     xml2_1.3.4                   </span></span>
<span><span class="co">##  [65] utf8_1.2.3                    XVector_0.40.0               </span></span>
<span><span class="co">##  [67] BiocVersion_3.17.1            pillar_1.9.0                 </span></span>
<span><span class="co">##  [69] stringr_1.5.0                 yulab.utils_0.0.6            </span></span>
<span><span class="co">##  [71] later_1.3.1                   dplyr_1.1.2                  </span></span>
<span><span class="co">##  [73] lattice_0.21-8                rtracklayer_1.60.0           </span></span>
<span><span class="co">##  [75] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [77] Biostrings_2.68.1             knitr_1.43                   </span></span>
<span><span class="co">##  [79] ProtGenerics_1.32.0           xfun_0.39                    </span></span>
<span><span class="co">##  [81] stringi_1.7.12                lazyeval_0.2.2               </span></span>
<span><span class="co">##  [83] yaml_2.3.7                    evaluate_0.21                </span></span>
<span><span class="co">##  [85] codetools_0.2-19              tibble_3.2.1                 </span></span>
<span><span class="co">##  [87] BiocManager_1.30.21           ggplotify_0.1.0              </span></span>
<span><span class="co">##  [89] cli_3.6.1                     xtable_1.8-4                 </span></span>
<span><span class="co">##  [91] systemfonts_1.0.4             munsell_0.5.0                </span></span>
<span><span class="co">##  [93] jquerylib_0.1.4               Rcpp_1.0.10                  </span></span>
<span><span class="co">##  [95] png_0.1-8                     XML_3.99-0.14                </span></span>
<span><span class="co">##  [97] parallel_4.3.0                ellipsis_0.3.2               </span></span>
<span><span class="co">##  [99] pkgdown_2.0.7                 blob_1.2.4                   </span></span>
<span><span class="co">## [101] prettyunits_1.1.1             bitops_1.0-7                 </span></span>
<span><span class="co">## [103] scales_1.2.1                  crayon_1.5.2                 </span></span>
<span><span class="co">## [105] RcppHMM_1.2.2                 rlang_1.1.1                  </span></span>
<span><span class="co">## [107] KEGGREST_1.40.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode_exclude" class="csl-entry">
Amemiya, Haley M, Anshul Kundaje, and Alan P Boyle. 2019. <span>“The
<span>ENCODE</span> Blacklist: Identification of Problematic Regions of
the Genome.”</span> <em>Scientific Reports</em> 9 (1): 9354. <a href="https://doi.org/10.1038/s41598-019-45839-z" class="external-link">https://doi.org/10.1038/s41598-019-45839-z</a>.
</div>
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy
R. Zhang. 2010. <span>“Subsampling Methods for Genomic
Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4):
1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-encode" class="csl-entry">
ENCODE. 2012. <span>“<span class="nocase">An integrated encyclopedia of
DNA elements in the human genome</span>.”</span> <em>Nature</em> 489
(7414): 57–74. <a href="https://doi.org/10.1038/nature11247" class="external-link">https://doi.org/10.1038/nature11247</a>.
</div>
<div id="ref-ensembl2021" class="csl-entry">
Howe, Kevin L, Premanand Achuthan, James Allen, Jamie Allen, Jorge
Alvarez-Jarreta, M Ridwan Amode, Irina M Armean, et al. 2020.
<span>“<span>Ensembl 2021</span>.”</span> <em>Nucleic Acids
Research</em> 49 (D1): D884–91. <a href="https://doi.org/10.1093/nar/gkaa942" class="external-link">https://doi.org/10.1093/nar/gkaa942</a>.
</div>
<div id="ref-Kramer2022" class="csl-entry">
Kramer, Nicole E, Eric S Davis, Craig D Wenger, Erika M Deoudes, Sarah M
Parker, Michael I Love, and Douglas H Phanstiel. 2022. <span>“<span class="nocase">Plotgardener: cultivating precise multi-panel figures in
R</span>.”</span> <em>Bioinformatics</em> 38 (7): 2042–45. <a href="https://doi.org/10.1093/bioinformatics/btac057" class="external-link">https://doi.org/10.1093/bioinformatics/btac057</a>.
</div>
<div id="ref-Lee2019" class="csl-entry">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges:
A Grammar of Genomic Data Transformation.”</span> <em>Genome
Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8" class="external-link">https://doi.org/10.1186/s13059-018-1597-8</a>.
</div>
<div id="ref-bootRanges" class="csl-entry">
Mu, Wancen, Eric S. Davis, Stuart Lee, Mikhail G. Dozmorov, Douglas H.
Phanstiel, and Michael I. Love. 2023. <span>“bootRanges: Flexible
Generation of Null Sets of Genomic Ranges for Hypothesis
Testing.”</span> <em>Bioinformatics</em>. <a href="https://doi.org/10.1093/bioinformatics/btad190" class="external-link">https://doi.org/10.1093/bioinformatics/btad190</a>.
</div>
<div id="ref-excluderanges" class="csl-entry">
Ogata, Jonathan D., Wancen Mu, Eric S. Davis, Bingjie Xue, J. Chuck
Harrell, Nathan C. Sheffield, Douglas H. Phanstiel, Michael I. Love, and
Mikhail G. Dozmorov. 2023. <span>“Excluderanges: Exclusion Sets for
T2T-CHM13, GRCm39, and Other Genome Assemblies.”</span>
<em>Bioinformatics</em>. <a href="https://doi.org/10.1093/bioinformatics/btad198" class="external-link">https://doi.org/10.1093/bioinformatics/btad198</a>.
</div>
<div id="ref-dnacopy" class="csl-entry">
Olshen, A. B., E. S. Venkatraman, R. Lucito, and M. Wigler. 2004.
<span>“<span class="nocase"><span>C</span>ircular binary segmentation
for the analysis of array-based
<span>D</span><span>N</span><span>A</span> copy number
data</span>.”</span> <em>Biostatistics</em> 5 (4): 557–72.
</div>
<div id="ref-ensembldb" class="csl-entry">
Rainer, Johannes, Laurent Gatto, and Christian X Weichenberger. 2019.
<span>“<span class="nocase">ensembldb: an R package to create and use
Ensembl-based annotation resources</span>.”</span>
<em>Bioinformatics</em> 35 (17): 3151–53. <a href="https://doi.org/10.1093/bioinformatics/btz031" class="external-link">https://doi.org/10.1093/bioinformatics/btz031</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://twitter.com/wancenm" class="external-link">Wancen Mu</a>, <a href="https://ericscottdavis.com" class="external-link">Eric Davis</a>, <a href="http://phanstiel-lab.med.unc.edu" class="external-link">Douglas Phanstiel</a>, <a href="https://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://chanzuckerberg.com" class="external-link">CZI</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
