<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Matching case study II: CTCF orientation • nullranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Matching case study II: CTCF orientation">
<meta property="og:description" content="nullranges">
<meta property="og:image" content="https://nullranges.github.io/nullranges/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/nullranges.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bootRanges.html">Introduction to bootRanges</a>
    </li>
    <li>
      <a href="../articles/matchRanges.html">Introduction to matchRanges</a>
    </li>
    <li>
      <a href="../articles/matching_granges.html">Matching case study I: CTCF occupancy</a>
    </li>
    <li>
      <a href="../articles/matching_ginteractions.html">Matching case study II: CTCF orientation</a>
    </li>
    <li>
      <a href="../articles/matching_pool_set.html">Creating a pool set for matchRanges</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nullranges/nullranges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Matching case study II: CTCF orientation</h1>
                        <h4 data-toc-skip class="author">Eric S.
Davis</h4>
            
            <h4 data-toc-skip class="date">05/17/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nullranges/nullranges/blob/HEAD/vignettes/matching_ginteractions.Rmd" class="external-link"><code>vignettes/matching_ginteractions.Rmd</code></a></small>
      <div class="hidden name"><code>matching_ginteractions.Rmd</code></div>

    </div>

    
    
<p>In this vignette, we demonstrate the generation of covariate-matched
null ranges by using the <code><a href="../reference/matchRanges.html">matchRanges()</a></code> function to test the
“covergence rule” of CTCF-bound chromatin loops, first described in Rao
et al. 2014.</p>
<div class="section level3">
<h3 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p>In the human genome, chromatin loops play an important role in
regulating gene expression by connecting regulatory loci to gene
promoters. The anchors of these loops are bound by the protein CTCF,
which is required for loop formation and maintenance. CTCF binds a
specific non-palindromic DNA sequence motif. And the direction of this
motif at loop anchors is non-random. That is, CTCF binding motifs found
at the anchors of loops tend to be oriented towards the center of the
loop more often then would be expected by chance. This phenomenon, first
described by Rao et al in 2014, is known as the “convergence rule”. It
was initially discovered by first filtering for loops that had a single
CTCF binding site at each end of the loop. Here we use matchRanges to
reinvestigate the convergence using all loops.</p>
<p>We will use <code>hg19_10kb_ctcfBoundBinPairs</code> data from the
<code>nullrangesData</code> package which contains features from the
GM12878 cell line aligned to hg19 as well as CTCF motif data from the <a href="https://bioconductor.org/packages/devel/data/annotation/html/CTCF.html" class="external-link">CTCF
data package</a> available on Bioconductor.
<code>hg19_10kb_ctcfBoundBinPairs</code> is a <a href="https://bioconductor.org/packages/release/bioc/html/InteractionSet.html" class="external-link"><code>GInteractions</code></a>
object with all pairwise combinations of CTCF-bound 10Kb bins within 1Mb
annotated with the following features:</p>
<ul>
<li>The total CTCF signal in each bin.</li>
<li>The number of CTCF sites in each bin.</li>
<li>The distance between bin pairs.</li>
<li>Whether at least one CTCF site is convergent between each bin
pair.</li>
<li>The presence or absence of a loop between each bin pair.</li>
</ul>
<p>Using these annotations and the <code><a href="../reference/matchRanges.html">matchRanges()</a></code> function,
we can compare CTCF motif orientations between pairs of genomic regions
that are 1) connected by loops, 2) not connected by loops, 3) randomly
chosen, or 4) not connected by loops, but matched for the strength of
CTCF sites and distance between loop anchors.</p>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-1-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="matching-with-matchranges">Matching with <code>matchRanges()</code><a class="anchor" aria-label="anchor" href="#matching-with-matchranges"></a>
</h3>
<p>Before we generate our null ranges, let’s take a look at our example
dataset:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span></span>
<span><span class="co">## Load example data</span></span>
<span><span class="va">binPairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/hg19_10kb_ctcfBoundBinPairs.html" class="external-link">hg19_10kb_ctcfBoundBinPairs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">binPairs</span></span></code></pre></div>
<pre><code><span><span class="co">## StrictGInteractions object with 198120 interactions and 5 metadata columns:</span></span>
<span><span class="co">##            seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">##                &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">##        [1]      chr1       230001-240000 ---      chr1       520001-530000 |</span></span>
<span><span class="co">##        [2]      chr1       230001-240000 ---      chr1       710001-720000 |</span></span>
<span><span class="co">##        [3]      chr1       230001-240000 ---      chr1       800001-810000 |</span></span>
<span><span class="co">##        [4]      chr1       230001-240000 ---      chr1       840001-850000 |</span></span>
<span><span class="co">##        [5]      chr1       230001-240000 ---      chr1       870001-880000 |</span></span>
<span><span class="co">##        ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">##   [198116]      chrX 154310001-154320000 ---      chrX 154370001-154380000 |</span></span>
<span><span class="co">##   [198117]      chrX 154310001-154320000 ---      chrX 155250001-155260000 |</span></span>
<span><span class="co">##   [198118]      chrX 154320001-154330000 ---      chrX 154370001-154380000 |</span></span>
<span><span class="co">##   [198119]      chrX 154320001-154330000 ---      chrX 155250001-155260000 |</span></span>
<span><span class="co">##   [198120]      chrX 154370001-154380000 ---      chrX 155250001-155260000 |</span></span>
<span><span class="co">##               looped ctcfSignal  n_sites  distance convergent</span></span>
<span><span class="co">##            &lt;logical&gt;  &lt;numeric&gt; &lt;factor&gt; &lt;integer&gt;  &lt;logical&gt;</span></span>
<span><span class="co">##        [1]     FALSE    5.18038        2    290000      FALSE</span></span>
<span><span class="co">##        [2]     FALSE    5.46775        2    480000       TRUE</span></span>
<span><span class="co">##        [3]     FALSE    7.30942        2    570000      FALSE</span></span>
<span><span class="co">##        [4]     FALSE    7.34338        2    610000      FALSE</span></span>
<span><span class="co">##        [5]     FALSE    6.31338        3    640000       TRUE</span></span>
<span><span class="co">##        ...       ...        ...      ...       ...        ...</span></span>
<span><span class="co">##   [198116]     FALSE    6.79246        2     60000      FALSE</span></span>
<span><span class="co">##   [198117]     FALSE    6.12447        3    940000       TRUE</span></span>
<span><span class="co">##   [198118]     FALSE    7.40868        2     50000       TRUE</span></span>
<span><span class="co">##   [198119]     FALSE    7.00936        3    930000      FALSE</span></span>
<span><span class="co">##   [198120]     FALSE    6.73402        3    880000       TRUE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   regions: 20612 ranges and 5 metadata columns</span></span>
<span><span class="co">##   seqinfo: 23 sequences from hg19 genome</span></span></code></pre>
<p>Let’s start by defining our focal set (i.e. looped bin-pairs), our
pool set (i.e un-looped bin-pairs), and our covariates of interest
(i.e. <code>ctcfSignal</code> and <code>distance</code>):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges">nullranges</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">mgi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/matchRanges.html">matchRanges</a></span><span class="op">(</span>focal <span class="op">=</span> <span class="va">binPairs</span><span class="op">[</span><span class="va">binPairs</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   pool <span class="op">=</span> <span class="va">binPairs</span><span class="op">[</span><span class="op">!</span><span class="va">binPairs</span><span class="op">$</span><span class="va">looped</span><span class="op">]</span>,</span>
<span>                   covar <span class="op">=</span> <span class="op">~</span><span class="va">ctcfSignal</span> <span class="op">+</span> <span class="va">distance</span> <span class="op">+</span> <span class="va">n_sites</span>,</span>
<span>                   method <span class="op">=</span> <span class="st">'stratified'</span><span class="op">)</span></span>
<span><span class="va">mgi</span></span></code></pre></div>
<pre><code><span><span class="co">## MatchedGInteractions object with 3104 interactions and 5 metadata columns:</span></span>
<span><span class="co">##          seqnames1             ranges1     seqnames2             ranges2 |</span></span>
<span><span class="co">##              &lt;Rle&gt;           &lt;IRanges&gt;         &lt;Rle&gt;           &lt;IRanges&gt; |</span></span>
<span><span class="co">##      [1]     chr11   62160001-62170000 ---     chr11   62190001-62200000 |</span></span>
<span><span class="co">##      [2]     chr17     7890001-7900000 ---     chr17     7990001-8000000 |</span></span>
<span><span class="co">##      [3]     chr22   36460001-36470000 ---     chr22   36680001-36690000 |</span></span>
<span><span class="co">##      [4]     chr11     1560001-1570000 ---     chr11     1710001-1720000 |</span></span>
<span><span class="co">##      [5]     chr12   31780001-31790000 ---     chr12   31830001-31840000 |</span></span>
<span><span class="co">##      ...       ...                 ... ...       ...                 ... .</span></span>
<span><span class="co">##   [3100]      chr7   25100001-25110000 ---      chr7   25220001-25230000 |</span></span>
<span><span class="co">##   [3101]      chr9   33850001-33860000 ---      chr9   34400001-34410000 |</span></span>
<span><span class="co">##   [3102]     chr11     6010001-6020000 ---     chr11     6300001-6310000 |</span></span>
<span><span class="co">##   [3103]     chr14 100750001-100760000 ---     chr14 100850001-100860000 |</span></span>
<span><span class="co">##   [3104]     chr11     1840001-1850000 ---     chr11     2170001-2180000 |</span></span>
<span><span class="co">##             looped ctcfSignal  n_sites  distance convergent</span></span>
<span><span class="co">##          &lt;logical&gt;  &lt;numeric&gt; &lt;factor&gt; &lt;integer&gt;  &lt;logical&gt;</span></span>
<span><span class="co">##      [1]     FALSE    8.12860        2     30000      FALSE</span></span>
<span><span class="co">##      [2]     FALSE    8.72976        3    100000       TRUE</span></span>
<span><span class="co">##      [3]     FALSE    9.51410        4    220000       TRUE</span></span>
<span><span class="co">##      [4]     FALSE    8.64759        3    150000      FALSE</span></span>
<span><span class="co">##      [5]     FALSE    8.43412        2     50000       TRUE</span></span>
<span><span class="co">##      ...       ...        ...      ...       ...        ...</span></span>
<span><span class="co">##   [3100]     FALSE    7.87730        2    120000      FALSE</span></span>
<span><span class="co">##   [3101]     FALSE    9.25850        3    550000       TRUE</span></span>
<span><span class="co">##   [3102]     FALSE    7.05064        2    290000      FALSE</span></span>
<span><span class="co">##   [3103]     FALSE    8.52685        4    100000       TRUE</span></span>
<span><span class="co">##   [3104]     FALSE    8.39022        3    330000       TRUE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   regions: 20612 ranges and 5 metadata columns</span></span>
<span><span class="co">##   seqinfo: 23 sequences from hg19 genome</span></span></code></pre>
<p>When the focal and pool arguments are <code>GInteractions</code>
objects, <code><a href="../reference/matchRanges.html">matchRanges()</a></code> returns a
<code>MatchedGInteractions</code> object. The
<code>MatchedGInteractions</code> class extends
<code>GInteractions</code> so all of the same operations can be
applied:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">## Summarize ctcfSignal by n_sites</span></span>
<span><span class="va">mgi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html" class="external-link">regions</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">n_sites</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>ctcfSignal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n_sites</span>, y <span class="op">=</span> <span class="va">ctcfSignal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, stroke <span class="op">=</span> <span class="fl">1</span>,  fill <span class="op">=</span> <span class="st">'white'</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>                                      fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-4-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="assessing-quality-of-matching">Assessing quality of matching<a class="anchor" aria-label="anchor" href="#assessing-quality-of-matching"></a>
</h3>
<p>We can get a quick summary of the matching quality with
<code><a href="../reference/overview.html">overview()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overview.html">overview</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span></span>
<span><span class="va">ov</span></span></code></pre></div>
<pre><code><span><span class="co">## MatchedGInteractions object: </span></span>
<span><span class="co">##        set      N ctcfSignal.mean ctcfSignal.sd distance.mean distance.sd</span></span>
<span><span class="co">##      focal   3104             8.3          0.67        320000      230000</span></span>
<span><span class="co">##    matched   3104             8.3          0.70        330000      250000</span></span>
<span><span class="co">##       pool 195016             7.9          0.85        490000      290000</span></span>
<span><span class="co">##  unmatched 191912             7.8          0.85        490000      290000</span></span>
<span><span class="co">##  n_sites.0 n_sites.1 n_sites.2 n_sites.3 n_sites.4 n_sites.5 n_sites.6</span></span>
<span><span class="co">##          0         0      2167       734       164        32         4</span></span>
<span><span class="co">##          0         0      2208       721       142        27         5</span></span>
<span><span class="co">##          0         0    152318     34992      5971       944       158</span></span>
<span><span class="co">##          0         0    150110     34271      5829       917       153</span></span>
<span><span class="co">##  n_sites.7+ ps.mean ps.sd</span></span>
<span><span class="co">##           3   0.026 0.016</span></span>
<span><span class="co">##           1   0.026 0.016</span></span>
<span><span class="co">##         633   0.016 0.013</span></span>
<span><span class="co">##         632   0.015 0.013</span></span>
<span><span class="co">## --------</span></span>
<span><span class="co">## focal - matched: </span></span>
<span><span class="co">##  ctcfSignal.mean ctcfSignal.sd distance.mean distance.sd n_sites.0 n_sites.1</span></span>
<span><span class="co">##          0.00038        -0.031         -2400      -27000         0         0</span></span>
<span><span class="co">##  n_sites.2 n_sites.3 n_sites.4 n_sites.5 n_sites.6 n_sites.7+  ps.mean    ps.sd</span></span>
<span><span class="co">##        -41        13        22         5        -1          2 -7.1e-08 -4.3e-07</span></span></code></pre>
<p>In addition to providing a printed overview, the overview data can be
extracted for convenience. For example, the <code>quality</code>
property shows the absolute value of the mean difference between focal
and matched sets. Therefore, the lower this value, the better the
matching quality:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ov</span><span class="op">$</span><span class="va">quality</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7.1e-08</span></span></code></pre>
<div class="section level4">
<h4 id="visualizing-matching-results">Visualizing matching results<a class="anchor" aria-label="anchor" href="#visualizing-matching-results"></a>
</h4>
<p>Let’s visualize overall matching quality by plotting propensity
scores for the focal, pool, and matched sets:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPropensity.html">plotPropensity</a></span><span class="op">(</span><span class="va">mgi</span>, sets <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'p'</span>, <span class="st">'m'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-7-1.png"><!-- --></p>
<p>Log transformations can be applied to ‘x’, ‘y’, or both
(<code>c('x', 'y')</code>) for plotting functions to make it easier to
assess quality. It is clear that the matched set is very well matched to
the focal set:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPropensity.html">plotPropensity</a></span><span class="op">(</span><span class="va">mgi</span>, sets <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'p'</span>, <span class="st">'m'</span><span class="op">)</span>, log <span class="op">=</span> <span class="st">'x'</span><span class="op">)</span></span></code></pre></div>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-8-1.png"><!-- --></p>
<p>We can ensure that covariate distributions have been matched
appropriately by using the <code><a href="../reference/covariates.html">covariates()</a></code> function to extract
matched covariates along with <code>patchwork</code> and
<code>plotCovarite</code> to visualize all distributions:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="../reference/covariates.html">covariates</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span>, <span class="va">plotCovariate</span>, x<span class="op">=</span><span class="va">mgi</span>, sets <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'f'</span>, <span class="st">'m'</span>, <span class="st">'p'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">'/'</span>, <span class="va">plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-9-1.png"><!-- --></p>
</div>
</div>
<div class="section level3">
<h3 id="compare-ctcf-site-orientation">Compare CTCF site orientation<a class="anchor" aria-label="anchor" href="#compare-ctcf-site-orientation"></a>
</h3>
<p>Using our matched ranges, we can compare the percent of looped pairs
with at least one convergent CTCF site against unlooped pairs, randomly
selected pairs, and pairs that are unlooped but have been matched for
our covariates. The accessor function <code><a href="../reference/focal.html">focal()</a></code> and
<code><a href="../reference/pool.html">pool()</a></code> can be used to conveniently extract these matched
sets:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Generate a randomly selected set from all binPairs</span></span>
<span><span class="va">all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="../reference/focal.html">focal</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span>, <span class="fu"><a href="../reference/pool.html">pool</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">random</span> <span class="op">&lt;-</span> <span class="va">all</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">all</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Calculate the percent of convergent CTCF sites for each group</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="../reference/focal.html">focal</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">$</span><span class="va">convergent</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/focal.html">focal</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="../reference/pool.html">pool</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">$</span><span class="va">convergent</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="../reference/pool.html">pool</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">random</span><span class="op">$</span><span class="va">convergent</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">random</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="va">g4</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">$</span><span class="va">convergent</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mgi</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="co">## Visualize</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span>height <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">g1</span>, <span class="va">g2</span>, <span class="va">g3</span>, <span class="va">g4</span><span class="op">)</span>,</span>
<span>        names.arg <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'looped\n(focal)'</span>, <span class="st">'unlooped\n(pool)'</span>,</span>
<span>                      <span class="st">'all pairs\n(random)'</span>, <span class="st">'unlooped\n(matched)'</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#1F78B4'</span>, <span class="st">'#33A02C'</span>, <span class="st">'orange2'</span>, <span class="st">'#A6CEE3'</span><span class="op">)</span>, </span>
<span>        ylab <span class="op">=</span> <span class="st">"Convergent CTCF Sites (%)"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Testing the Convergence Rule"</span>,</span>
<span>        border <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>        las <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="matching_ginteractions_files/figure-html/unnamed-chunk-10-1.png"><!-- --></p>
<p>As shown above the convergent rule holds true even when controlling
for CTCF signal strength and bin pair distance. Greater than 90% of
looped pairs contain convergent CTCF sites, compared to only ~55% among
non-looped subsets.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] patchwork_1.1.2             plyranges_1.20.0           </span></span>
<span><span class="co">##  [3] nullrangesData_1.6.0        ExperimentHub_2.8.0        </span></span>
<span><span class="co">##  [5] AnnotationHub_3.8.0         BiocFileCache_2.8.0        </span></span>
<span><span class="co">##  [7] dbplyr_2.3.2                ggplot2_3.4.2              </span></span>
<span><span class="co">##  [9] plotgardener_1.6.0          nullranges_1.7.0           </span></span>
<span><span class="co">## [11] InteractionSet_1.28.0       SummarizedExperiment_1.30.1</span></span>
<span><span class="co">## [13] Biobase_2.60.0              MatrixGenerics_1.12.0      </span></span>
<span><span class="co">## [15] matrixStats_0.63.0          GenomicRanges_1.52.0       </span></span>
<span><span class="co">## [17] GenomeInfoDb_1.36.0         IRanges_2.34.0             </span></span>
<span><span class="co">## [19] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3            strawr_0.0.91                </span></span>
<span><span class="co">##   [3] jsonlite_1.8.4                magrittr_2.0.3               </span></span>
<span><span class="co">##   [5] farver_2.1.1                  rmarkdown_2.21               </span></span>
<span><span class="co">##   [7] fs_1.6.2                      BiocIO_1.10.0                </span></span>
<span><span class="co">##   [9] zlibbioc_1.46.0               ragg_1.2.5                   </span></span>
<span><span class="co">##  [11] vctrs_0.6.2                   memoise_2.0.1                </span></span>
<span><span class="co">##  [13] Rsamtools_2.16.0              RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [15] progress_1.2.2                htmltools_0.5.5              </span></span>
<span><span class="co">##  [17] S4Arrays_1.0.4                curl_5.0.0                   </span></span>
<span><span class="co">##  [19] gridGraphics_0.5-1            sass_0.4.6                   </span></span>
<span><span class="co">##  [21] pracma_2.4.2                  KernSmooth_2.23-21           </span></span>
<span><span class="co">##  [23] bslib_0.4.2                   desc_1.4.2                   </span></span>
<span><span class="co">##  [25] cachem_1.0.8                  GenomicAlignments_1.36.0     </span></span>
<span><span class="co">##  [27] mime_0.12                     lifecycle_1.0.3              </span></span>
<span><span class="co">##  [29] pkgconfig_2.0.3               Matrix_1.5-4                 </span></span>
<span><span class="co">##  [31] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [33] GenomeInfoDbData_1.2.10       shiny_1.7.4                  </span></span>
<span><span class="co">##  [35] digest_0.6.31                 colorspace_2.1-0             </span></span>
<span><span class="co">##  [37] AnnotationDbi_1.62.1          rprojroot_2.0.3              </span></span>
<span><span class="co">##  [39] textshaping_0.3.6             RSQLite_2.3.1                </span></span>
<span><span class="co">##  [41] filelock_1.0.2                labeling_0.4.2               </span></span>
<span><span class="co">##  [43] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">##  [45] compiler_4.3.0                bit64_4.0.5                  </span></span>
<span><span class="co">##  [47] withr_2.5.0                   BiocParallel_1.34.1          </span></span>
<span><span class="co">##  [49] DBI_1.1.3                     highr_0.10                   </span></span>
<span><span class="co">##  [51] rappdirs_0.3.3                DelayedArray_0.26.2          </span></span>
<span><span class="co">##  [53] rjson_0.2.21                  tools_4.3.0                  </span></span>
<span><span class="co">##  [55] interactiveDisplayBase_1.38.0 httpuv_1.6.11                </span></span>
<span><span class="co">##  [57] glue_1.6.2                    restfulr_0.0.15              </span></span>
<span><span class="co">##  [59] promises_1.2.0.1              generics_0.1.3               </span></span>
<span><span class="co">##  [61] gtable_0.3.3                  hms_1.1.3                    </span></span>
<span><span class="co">##  [63] data.table_1.14.8             utf8_1.2.3                   </span></span>
<span><span class="co">##  [65] XVector_0.40.0                BiocVersion_3.17.1           </span></span>
<span><span class="co">##  [67] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">##  [69] yulab.utils_0.0.6             later_1.3.1                  </span></span>
<span><span class="co">##  [71] dplyr_1.1.2                   lattice_0.21-8               </span></span>
<span><span class="co">##  [73] rtracklayer_1.60.0            bit_4.0.5                    </span></span>
<span><span class="co">##  [75] ks_1.14.0                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [77] Biostrings_2.68.0             knitr_1.42                   </span></span>
<span><span class="co">##  [79] xfun_0.39                     stringi_1.7.12               </span></span>
<span><span class="co">##  [81] yaml_2.3.7                    evaluate_0.21                </span></span>
<span><span class="co">##  [83] codetools_0.2-19              tibble_3.2.1                 </span></span>
<span><span class="co">##  [85] BiocManager_1.30.20           ggplotify_0.1.0              </span></span>
<span><span class="co">##  [87] cli_3.6.1                     xtable_1.8-4                 </span></span>
<span><span class="co">##  [89] systemfonts_1.0.4             munsell_0.5.0                </span></span>
<span><span class="co">##  [91] jquerylib_0.1.4               Rcpp_1.0.10                  </span></span>
<span><span class="co">##  [93] png_0.1-8                     XML_3.99-0.14                </span></span>
<span><span class="co">##  [95] parallel_4.3.0                ellipsis_0.3.2               </span></span>
<span><span class="co">##  [97] pkgdown_2.0.7                 blob_1.2.4                   </span></span>
<span><span class="co">##  [99] prettyunits_1.1.1             mclust_6.0.0                 </span></span>
<span><span class="co">## [101] bitops_1.0-7                  mvtnorm_1.1-3                </span></span>
<span><span class="co">## [103] scales_1.2.1                  ggridges_0.5.4               </span></span>
<span><span class="co">## [105] purrr_1.0.1                   crayon_1.5.2                 </span></span>
<span><span class="co">## [107] rlang_1.1.1                   KEGGREST_1.40.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://twitter.com/wancenm" class="external-link">Wancen Mu</a>, <a href="https://ericscottdavis.com" class="external-link">Eric Davis</a>, <a href="http://phanstiel-lab.med.unc.edu" class="external-link">Douglas Phanstiel</a>, <a href="https://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://chanzuckerberg.com" class="external-link">CZI</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
