<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to nullranges • nullranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to nullranges">
<meta property="og:description" content="nullranges">
<meta property="og:image" content="https://nullranges.github.io/nullranges/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/nullranges.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/matching_ranges.html">Overview of matchRanges</a>
    </li>
    <li>
      <a href="../articles/matching_granges.html">Case study I: CTCF occupancy</a>
    </li>
    <li>
      <a href="../articles/matching_ginteractions.html">Case study II: CTCF orientation</a>
    </li>
    <li>
      <a href="../articles/segmented_boot_ranges.html">Segmented block bootstrap</a>
    </li>
    <li>
      <a href="../articles/unseg_boot_ranges.html">Unsegmented block bootstrap</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nullranges/nullranges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="nullranges_files/htmlwidgets-1.6.0/htmlwidgets.js"></script><script src="nullranges_files/viz-1.8.2/viz.js"></script><link href="nullranges_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="nullranges_files/grViz-binding-1.0.9/grViz.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to nullranges</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nullranges/nullranges/blob/HEAD/vignettes/nullranges.Rmd" class="external-link"><code>vignettes/nullranges.Rmd</code></a></small>
      <div class="hidden name"><code>nullranges.Rmd</code></div>

    </div>

    
    
<p>The <em>nullranges</em> package contains functions for generation of
feature sets (genomic regions) for exploring the null hypothesis of
overlap or colocalization of two observed feature sets. This package
contains a number of vignettes describing different functionality, from
basic to more advanced usage. For a listing of all the vignettes in the
package, one can type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"nullranges"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="choice-of-methods">Choice of methods<a class="anchor" aria-label="anchor" href="#choice-of-methods"></a>
</h3>
<p>The <em>nullranges</em> package has two branches of functionality:
<em>matching</em> or <em>bootstrapping</em> to generate null feature
sets.</p>
<p>In the vignette sections below we describe these two branches and
give more formal definitions, where each branch has associated vignettes
and man page sections. To give a high level overview, we first provide a
<strong>decision tree</strong> that helps to indicate the considerations
when choosing between these branches of functionality.</p>
<p>Define <strong>features</strong> as a set of genomic locations of
interest, these are minimally a chromosome, start and width (1 bp or
more), and optionally strand and metadata. Given a set of features,
suppose that we wish to create an alternate set that represents a null
feature set, sometimes also referred to as “background” or a “control
set”. For example, we may see that our primary features are close to
gene start sites, but are they closer than we would expect compared to a
reasonable choice of null/background/control features?</p>
<p>Additionally, define <strong>pool</strong> as a much larger set of
genomic locations compared to the primary features, and
<strong>covariates</strong> as pieces of metadata attached to all of the
considered features (stored as <code>mcols(&lt;GRanges&gt;)</code>,
which may be continuous, integer, factor, etc.).</p>
<p>Our choice of methods is informed by the following:</p>
<ul>
<li>Are the features defined with respect to a pool, e.g. open chromatin
sites bound by a protein, defined with respect to all open chromatin in
a particular cell type?</li>
<li>Are there important, potentially confounding covariates, e.g. does
GC content, distance to particular landmarks, expression level, LD
score, etc. potentially explain distribution of features in a way that
should be controlled for in downstream analysis?</li>
<li>Do the primary feature set and pool have different distribution of
these potential confounding covariates?</li>
<li>Does the local genomic context matter (<em>local</em> defined in
genomic coordinate space)? Are we asking if primary features are
<em>near in the genome</em> to the gene start sites or some other
genomic feature set?</li>
<li>Do the features tend to be distributed in genomic coordinate space
such that they are clustered, e.g. open chromatin sites tend to occur
near each other? Or do features that are near each other in the genome
tend to have similar metadata of interest to the biological question,
e.g. CpG’s near each other having similar levels of methylation, which
will be used in downstream computations of average methylation near gene
start site?</li>
</ul>
<p>The following decision tree then informs what methods to choose:</p>
<div id="htmlwidget-dc5da67591ff86b20d24" style="width:528px;height:500px;" class="grViz html-widget "></div>
<script type="application/json" data-for="htmlwidget-dc5da67591ff86b20d24">{"x":{"diagram":"digraph {\n  graph [layout = dot, rankdir = TB]\n  \n  node [shape = rectangle]        \n  source [label = \"Features defined with respect to a pool\"]\n  poolyes [label = \"Yes\"]\n  poolno [label =  \"No\"]\n  cov [label = \"Potential confounding covariates\"]\n  covyes [label = \"Yes\"]\n  covno [label = \"No\"]\n  diffcov [label = \"Different covariate distribution\"]\n  diffcovyes [label = \"Yes\"]\n  diffcovno [label =  \"No\"]\n  match [label = \"matchRanges\"]\n  random [label = \"Random sample\"]\n  context [label = \"Local genomic context matters\"]\n  contextyes [label = \"Yes\"]\n  contextno [label = \"No\"]\n  cluster [label = \"Features cluster in genome\"]\n  random2 [label = \"Random sample\"]\n  clusteryes [label = \"Yes\"]\n  clusterno [label = \"No\"]\n  boot [label = \"bootRanges\"]\n  shuffle [label = \"Shuffle + exclusion\"]\n\n  source -> poolyes\n  source -> poolno\n  poolyes -> cov\n  cov -> covyes\n  cov -> covno\n  covyes -> diffcov\n  covno -> random\n  diffcov -> diffcovyes\n  diffcov -> diffcovno\n  diffcovyes -> match\n  diffcovno -> random\n  poolno -> context\n  context -> contextyes\n  context -> contextno\n  contextyes -> cluster\n  contextno -> random2\n  cluster -> clusteryes\n  cluster -> clusterno\n  clusteryes -> boot\n  clusterno -> shuffle\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="related-work">Related work<a class="anchor" aria-label="anchor" href="#related-work"></a>
</h3>
<p>For general considerations of generation of null feature sets or
segmentation for enrichment or colocalization analysis, consider the
papers of <span class="citation">De, Pedersen, and Kechris
(2014)</span>, <span class="citation">Haiminen, Mannila, and Terzi
(2007)</span>, <span class="citation">Huen and Russell (2010)</span>,
<span class="citation">Ferkingstad, Holden, and Sandve (2015)</span>,
<span class="citation">Dozmorov (2017)</span>, <span class="citation">Kanduri et al. (2019)</span> (with links in references
below). Other Bioconductor packages that offer randomization techniques
for enrichment analysis include <a href="https://bioconductor.org/packages/LOLA" class="external-link">LOLA</a> <span class="citation">(Sheffield and Bock 2016)</span> and <a href="https://bioconductor.org/packages/regioneR" class="external-link">regioneR</a> <span class="citation">(Gel et al. 2016)</span>. Methods implemented outside
of Bioconductor include <a href="https://github.com/AndreasHeger/gat" class="external-link">GAT</a> <span class="citation">(Heger et al. 2013)</span>, <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a> <span class="citation">(Bickel et al. 2010)</span>, <a href="http://bejerano.stanford.edu/great/public/html/" class="external-link">GREAT</a> <span class="citation">(McLean et al. 2010)</span>, <a href="https://github.com/favorov/GenometriCorr" class="external-link">GenometriCorr</a> <span class="citation">(Favorov et al. 2012)</span>, <a href="http://chip-enrich.med.umich.edu/" class="external-link">ChIP-Enrich</a> <span class="citation">(Welch et al. 2014)</span>, and <a href="https://dputhier.github.io/pygtftk/ologram.html" class="external-link">OLOGRAM</a> <span class="citation">(Ferré et al. 2019)</span>.</p>
<p>We note that our block bootstrapping approach closely follows that of
<a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a>, while offering
additional features/visualizations, and is re-implemented within
R/Bioconductor with efficient vectorized code for working with
<em>GRanges</em> objects <span class="citation">(Lawrence 2013)</span>.
In addition, <em>regioneR</em> and <em>OLOGRAM</em> also offer
randomization schemes that can preserve e.g. inter-feature
distances.</p>
<p>In the context of these related works, a design choice of
<em>nullranges</em> is its modularity: it solely generates null or
background feature sets as Bioconductor objects, such that these can be
used in workflows that compute and perform inference on arbitrary
statistics within the <em>plyranges</em> <span class="citation">(Stuart
Lee, Cook, and Lawrence 2019)</span> framework. We have referred to such
workflows as “fluent genomics” <span class="citation">(S. Lee, Lawrence,
and Love 2020)</span>, or “tidy genomics”, as these workflows build up
complex operations from commonly recognized verbs (filter, mutate, join,
group by, summarize, etc.) strung together with pipes
(<code>%&gt;%</code>).</p>
</div>
<div class="section level3">
<h3 id="further-description-of-matching-and-bootstrapping">Further description of matching and bootstrapping<a class="anchor" aria-label="anchor" href="#further-description-of-matching-and-bootstrapping"></a>
</h3>
<p>Suppose we want to examine the significance of overlaps of genomic
sets of features <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>. To test the significance of this
overlap, we calculate the overlap expected under the null by generating
a null feature set <span class="math inline">\(y'\)</span>
(potentially many times). The null features in <span class="math inline">\(y'\)</span> may be characterized by:</p>
<ol style="list-style-type: decimal">
<li>Drawing from a larger pool <span class="math inline">\(z\)</span>
(<span class="math inline">\(y' \subset z\)</span>), such that <span class="math inline">\(y\)</span> and <span class="math inline">\(y'\)</span> have a similar distribution over
one or more covariates. This is the “matching” case. Note that the
features in <span class="math inline">\(y'\)</span> are original
features, just drawn from a different pool than y. The
<em>matchRanges</em> method is described in <span class="citation">Davis
et al. (2022)</span> <a href="https://doi.org/10.1101/2022.08.05.502985" class="external-link">doi:
10.1101/2022.08.05.502985</a>.</li>
<li>Generating a new set of genomic features <span class="math inline">\(y'\)</span>, constructing them from the
original set <span class="math inline">\(y\)</span> by selecting blocks
of the genome with replacement, i.e. such that features can be sampled
more than once. This is the “bootstrapping” case. Note that, in this
case, <span class="math inline">\(y'\)</span> is an artificial
feature set, although the re-sampled features can retain covariates such
as score from the original feature set <span class="math inline">\(y\)</span>. The <em>bootRanges</em> method is
described in <span class="citation">Mu et al. (2022)</span> <a href="https://doi.org/10.1101/2022.09.02.506382" class="external-link">doi:
10.1101/2022.09.02.506382</a>.</li>
</ol>
</div>
<div class="section level3">
<h3 id="in-other-words">In other words<a class="anchor" aria-label="anchor" href="#in-other-words"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Matching – drawing from a pool of features but controlling for
certain characteristics, or <em>covariates</em>
</li>
<li>Bootstrapping – placing a number of artificial features in the
genome but controlling for the local dependence among features
(e.g. features clustering in the genome and/or having correlated
metadata)</li>
</ol>
</div>
<div class="section level3">
<h3 id="options-and-features">Options and features<a class="anchor" aria-label="anchor" href="#options-and-features"></a>
</h3>
<p>We provide a number of vignettes to describe the different matching
and bootstrapping use cases. In the matching case, we have implemented a
number of options, including nearest neighbor matching or rejection
sampling based matching. In the bootstrapping case, we have implemented
options for bootstrapping across or within chromosomes, and
bootstrapping only within states of a segmented genome. We also provide
a function to segment the genome by density of features. For example,
supposing that <span class="math inline">\(x\)</span> is a subset of
genes, we may want to generate <span class="math inline">\(y'\)</span> from <span class="math inline">\(y\)</span> such that features are re-sampled in
blocks from segments across the genome with similar gene density. In
both cases, we provide a number of functions for performing quality
control via visual inspection of diagnostic plots.</p>
</div>
<div class="section level3">
<h3 id="consideration-of-excluded-regions">Consideration of excluded regions<a class="anchor" aria-label="anchor" href="#consideration-of-excluded-regions"></a>
</h3>
<p>Finally, we recommend to incorporate list of regions where artificial
features should <em>not</em> be placed, including the ENCODE Exclusion
List <span class="citation">(Amemiya, Kundaje, and Boyle 2019)</span>.
This and other excluded ranges are made available in the <a href="https://dozmorovlab.github.io/excluderanges/" class="external-link">excluderanges</a>
Bioconductor package by Mikhail Dozmorov <em>et al.</em> <span class="citation">(Ogata et al. 2022)</span>. Use of excluded ranges is
demonstrated in the segmented block bootstrap vignette.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode_exclude" class="csl-entry">
Amemiya, Haley M, Anshul Kundaje, and Alan P Boyle. 2019. <span>“The
<span>ENCODE</span> Blacklist: Identification of Problematic Regions of
the Genome.”</span> <em>Scientific Reports</em> 9 (1): 9354. <a href="https://doi.org/10.1038/s41598-019-45839-z" class="external-link">https://doi.org/10.1038/s41598-019-45839-z</a>.
</div>
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy
R. Zhang. 2010. <span>“Subsampling Methods for Genomic
Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4):
1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-matchRanges" class="csl-entry">
Davis, Eric S., Wancen Mu, Stuart Lee, Mikhail G. Dozmorov, Michael I.
Love, and Douglas H. Phanstiel. 2022. <span>“matchRanges: Generating
Null Hypothesis Genomic Ranges via Covariate-Matched Sampling.”</span>
<em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.08.05.502985" class="external-link">https://doi.org/10.1101/2022.08.05.502985</a>.
</div>
<div id="ref-de_2014" class="csl-entry">
De, Subhajyoti, Brent S Pedersen, and Katerina Kechris. 2014. <span>“The
Dilemma of Choosing the Ideal Permutation Strategy While Estimating
Statistical Significance of Genome-Wide Enrichment.”</span>
<em>Briefings in Bioinformatics</em> 15 (6): 919–28. <a href="https://doi.org/10.1093/bib/bbt053" class="external-link">https://doi.org/10.1093/bib/bbt053</a>.
</div>
<div id="ref-dozmorov2017" class="csl-entry">
Dozmorov, Mikhail G. 2017. <span>“<span class="nocase">Epigenomic
annotation-based interpretation of genomic data: from enrichment
analysis to machine learning</span>.”</span> <em>Bioinformatics</em> 33
(20): 3323–30. <a href="https://doi.org/10.1093/bioinformatics/btx414" class="external-link">https://doi.org/10.1093/bioinformatics/btx414</a>.
</div>
<div id="ref-GenometriCorr" class="csl-entry">
Favorov, Alexander, Loris Mularoni, Leslie M Cope, Yulia Medvedeva,
Andrey A Mironov, Vsevolod J Makeev, and Sarah J Wheelan. 2012.
<span>“Exploring Massive, Genome Scale Datasets with the
<span>GenometriCorr</span> Package.”</span> <em><span>PLoS</span>
Computational Biology</em> 8 (5): e1002529. <a href="https://doi.org/10.1371/journal.pcbi.1002529" class="external-link">https://doi.org/10.1371/journal.pcbi.1002529</a>.
</div>
<div id="ref-ferkingstad2015" class="csl-entry">
Ferkingstad, Egil, Lars Holden, and Geir Kjetil Sandve. 2015.
<span>“<span class="nocase">Monte Carlo Null Models for Genomic
Data</span>.”</span> <em>Statistical Science</em> 30 (1): 59–71. <a href="https://doi.org/10.1214/14-STS484" class="external-link">https://doi.org/10.1214/14-STS484</a>.
</div>
<div id="ref-ologram" class="csl-entry">
Ferré, Q, G Charbonnier, N Sadouni, F Lopez, Y Kermezli, S Spicuglia, C
Capponi, B Ghattas, and D Puthier. 2019. <span>“<span class="nocase">OLOGRAM: determining significance of total overlap length
between genomic regions sets</span>.”</span> <em>Bioinformatics</em> 36
(6): 1920–22. <a href="https://doi.org/10.1093/bioinformatics/btz810" class="external-link">https://doi.org/10.1093/bioinformatics/btz810</a>.
</div>
<div id="ref-regioneR" class="csl-entry">
Gel, Bernat, Anna Díez-Villanueva, Eduard Serra, Marcus Buschbeck,
Miguel A Peinado, and Roberto Malinverni. 2016. <span>“<span class="nocase">regioneR</span>: An r/Bioconductor Package for the
Association Analysis of Genomic Regions Based on Permutation
Tests.”</span> <em>Bioinformatics</em> 32 (2): 289–91. <a href="https://doi.org/10.1093/bioinformatics/btv562" class="external-link">https://doi.org/10.1093/bioinformatics/btv562</a>.
</div>
<div id="ref-haiminen_2007" class="csl-entry">
Haiminen, Niina, Heikki Mannila, and Evimaria Terzi. 2007.
<span>“Comparing Segmentations by Applying Randomization
Techniques.”</span> <em><span>BMC</span> Bioinformatics</em> 8 (May):
171. <a href="https://doi.org/10.1186/1471-2105-8-171" class="external-link">https://doi.org/10.1186/1471-2105-8-171</a>.
</div>
<div id="ref-GAT" class="csl-entry">
Heger, Andreas, Caleb Webber, Martin Goodson, Chris P Ponting, and
Gerton Lunter. 2013. <span>“<span>GAT</span>: A Simulation Framework for
Testing the Association of Genomic Intervals.”</span>
<em>Bioinformatics</em> 29 (16): 2046–48. <a href="https://doi.org/10.1093/bioinformatics/btt343" class="external-link">https://doi.org/10.1093/bioinformatics/btt343</a>.
</div>
<div id="ref-huen_2010" class="csl-entry">
Huen, David S, and Steven Russell. 2010. <span>“On the Use of Resampling
Tests for Evaluating Statistical Significance of Binding-Site
Co-Occurrence.”</span> <em><span>BMC</span> Bioinformatics</em> 11
(June): 359. <a href="https://doi.org/10.1186/1471-2105-11-359" class="external-link">https://doi.org/10.1186/1471-2105-11-359</a>.
</div>
<div id="ref-kanduri_2019" class="csl-entry">
Kanduri, Chakravarthi, Christoph Bock, Sveinung Gundersen, Eivind Hovig,
and Geir Kjetil Sandve. 2019. <span>“Colocalization Analyses of Genomic
Elements: Approaches, Recommendations and Challenges.”</span>
<em>Bioinformatics</em> 35 (9): 1615–24. <a href="https://doi.org/10.1093/bioinformatics/bty835" class="external-link">https://doi.org/10.1093/bioinformatics/bty835</a>.
</div>
<div id="ref-granges" class="csl-entry">
Lawrence, Wolfgang AND Pagès, Michael AND Huber. 2013. <span>“Software
for Computing and Annotating Genomic Ranges.”</span> <em>PLOS
Computational Biology</em> 9 (8): 1–10. <a href="https://doi.org/10.1371/journal.pcbi.1003118" class="external-link">https://doi.org/10.1371/journal.pcbi.1003118</a>.
</div>
<div id="ref-fluent" class="csl-entry">
Lee, S, M Lawrence, and MI Love. 2020. <span>“Fluent Genomics
with Plyranges and Tximeta [Version 1; Peer Review: 1 Approved, 2
Approved with Reservations].”</span> <em>F1000Research</em> 9 (109). <a href="https://doi.org/10.12688/f1000research.22259.1" class="external-link">https://doi.org/10.12688/f1000research.22259.1</a>.
</div>
<div id="ref-Lee2019" class="csl-entry">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges:
A Grammar of Genomic Data Transformation.”</span> <em>Genome
Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8" class="external-link">https://doi.org/10.1186/s13059-018-1597-8</a>.
</div>
<div id="ref-GREAT" class="csl-entry">
McLean, Cory Y., Dave Bristor, Michael Hiller, Shoa L. Clarke, Bruce T.
Schaar, Craig B. Lowe, Aaron M. Wenger, and Gill Bejerano. 2010.
<span>“GREAT Improves Functional Interpretation of Cis-Regulatory
Regions.”</span> <em>Nature Biotechnology</em> 28 (5): 495–501. <a href="https://doi.org/10.1038/nbt.1630" class="external-link">https://doi.org/10.1038/nbt.1630</a>.
</div>
<div id="ref-bootRanges" class="csl-entry">
Mu, Wancen, Eric S. Davis, Stuart Lee, Mikhail G. Dozmorov, Douglas H.
Phanstiel, and Michael I. Love. 2022. <span>“bootRanges: Flexible
Generation of Null Sets of Genomic Ranges for Hypothesis
Testing.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.09.02.506382" class="external-link">https://doi.org/10.1101/2022.09.02.506382</a>.
</div>
<div id="ref-excluderanges" class="csl-entry">
Ogata, Jonathan D., Wancen Mu, Eric S. Davis, Bingjie Xue, J. Chuck
Harrell, Nathan C. Sheffield, Douglas H. Phanstiel, Michael I. Love, and
Mikhail G. Dozmorov. 2022. <span>“Excluderanges: Exclusion Sets for
T2T-CHM13, GRCm39, and Other Genome Assemblies.”</span>
<em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.11.21.517407" class="external-link">https://doi.org/10.1101/2022.11.21.517407</a>.
</div>
<div id="ref-LOLA" class="csl-entry">
Sheffield, Nathan C, and Christoph Bock. 2016. <span>“<span>LOLA</span>:
Enrichment Analysis for Genomic Region Sets and Regulatory Elements in r
and Bioconductor.”</span> <em>Bioinformatics</em> 32 (4): 587–89. <a href="https://doi.org/10.1093/bioinformatics/btv612" class="external-link">https://doi.org/10.1093/bioinformatics/btv612</a>.
</div>
<div id="ref-ChIP-Enrich" class="csl-entry">
Welch, Ryan P., Chee Lee, Paul M. Imbriano, Snehal Patil, Terry E.
Weymouth, R. Alex Smith, Laura J. Scott, and Maureen A. Sartor. 2014.
<span>“<span class="nocase">ChIP-Enrich: gene set enrichment testing for
ChIP-seq data</span>.”</span> <em>Nucleic Acids Research</em> 42 (13):
e105–5. <a href="https://doi.org/10.1093/nar/gku463" class="external-link">https://doi.org/10.1093/nar/gku463</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://twitter.com/wancenm" class="external-link">Wancen Mu</a>, <a href="https://ericscottdavis.com" class="external-link">Eric Davis</a>, <a href="http://phanstiel-lab.med.unc.edu" class="external-link">Douglas Phanstiel</a>, <a href="https://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://chanzuckerberg.com" class="external-link">CZI</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
