<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Segmented block bootstrap • nullranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Segmented block bootstrap">
<meta property="og:description" content="nullranges">
<meta property="og:image" content="https://nullranges.github.io/nullranges/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/nullranges.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/matching_ranges.html">Overview of matchRanges</a>
    </li>
    <li>
      <a href="../articles/matching_granges.html">Case study I: CTCF occupancy</a>
    </li>
    <li>
      <a href="../articles/matching_ginteractions.html">Case study II: CTCF orientation</a>
    </li>
    <li>
      <a href="../articles/segmented_boot_ranges.html">Segmented block bootstrap</a>
    </li>
    <li>
      <a href="../articles/unseg_boot_ranges.html">Unsegmented block bootstrap</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nullranges/nullranges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Segmented block bootstrap</h1>
                        <h4 data-toc-skip class="author">Wancen Mu</h4>
            
            <h4 data-toc-skip class="date">08/03/2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nullranges/nullranges/blob/HEAD/vignettes/segmented_boot_ranges.Rmd" class="external-link"><code>vignettes/segmented_boot_ranges.Rmd</code></a></small>
      <div class="hidden name"><code>segmented_boot_ranges.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The following vignette describes the <em>nullranges</em>
implementation of the block bootstrap with respect to a genomic
segmentation. See the main nullranges vignette for an overview of the
idea of bootstrapping, or below diagram, and there is additionally a
vignette on block boostrapping without respect to segmentation-<a href="unseg_boot_ranges.html">Unsegmented block bootstrap</a>.</p>
<p>As proposed by <span class="citation">Bickel et al. (2010)</span>,
<em>nullranges</em> contains an implementation of a block bootstrap,
such that features are sampled from the genome in blocks. The original
block bootstrapping algorithm is implemented in a python software called
Genome Structure Correlation, <a href="https://github.com/ParkerLab/encodegsc" class="external-link">GSC</a>.</p>
<p>In a segmented block bootstrap, the blocks are sampled and placed
within regions of a genome <em>segmentation</em>. That is, for a genome
segmented into states 1,2,…,S, blocks from state <em>s</em> will be used
to tile the ranges of state <em>s</em> in each bootstrap sample. The
process can be visualized in (A), a block with length <span class="math inline">\(L_b\)</span> is randomly selected from state “red”
and move to a tile block across chromosome. Additionally, the workflow
of bootRanges is diagrammed in (B) and listed as:</p>
<ol style="list-style-type: decimal">
<li>Overlap <em>GRanges</em> of feature <span class="math inline">\(x\)</span> and <em>GRanges</em> of feature <span class="math inline">\(y\)</span> to derive interested observed
statistics</li>
<li>
<code><a href="../reference/bootRanges.html">bootRanges()</a></code> with optional <code>segmentation</code>
and <code>exclude</code> creates a <em>bootRanges</em> object <span class="math inline">\(y'\)</span>
</li>
<li>Overlap <em>GRanges</em> of feature <span class="math inline">\(x\)</span> and <span class="math inline">\(y'\)</span> to derive interested bootstrap
statistics</li>
<li>
<span class="math inline">\(z\)</span> test is performed for testing
the null hypothesis that there is no true biological enrichment</li>
</ol>
<p><img src="images/bootRanges.jpg"><!-- --></p>
<p>The segmented block bootstrap has two options, either:</p>
<ul>
<li>Perform a de-novo segmentation of the genome using feature density,
e.g. gene density</li>
<li>Use exiting segmentation (e.g. ChromHMM, etc.) downloaded from
AnnotationHub or external to Bioconductor (BED files imported with
<em>rtracklayer</em>)</li>
</ul>
<p>In this vignette, we give an example of segmenting the hg38 genome by
Ensembl gene density, create bootstrapped peaks and evaluate overlaps
for observed peaks and bootstrap peaks, then we profile the time to
generate a single block bootstrap sample. Finally, we use a toy dataset
to visualize what a segmented block bootstrap sample looks like with
respect to a genome segmentation.</p>
<p>A finally consideration is whether the blocks should scale
proportionally to the segment state length, with the default setting of
<code>proportionLength=TRUE</code>. When blocks scale proportionally,
<code>blockLength</code> provides the maximal length of a block, while
the actual block length used for a segmentation state is proportional to
the fraction of genomic basepairs covered by that state. This option is
visualized on toy data at the end of this vignette.</p>
</div>
<div class="section level2">
<h2 id="pre-built-segmentations">Pre-built segmentations<a class="anchor" aria-label="anchor" href="#pre-built-segmentations"></a>
</h2>
<p><em>nullranges</em> has generated pre-built segmentations for easy
use by following below section <a href="#Segmentation%20by%20gene%20density">Segmentation by gene
density</a>. Either pre-built segmentations using <em>CBS</em> or
<em>HMM</em> methods with <span class="math inline">\(L_s=2e6\)</span>
considering excludable regions can be selected from
<em>ExperimentHub</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span><span class="op">)</span>
<span class="va">eh</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2022-07-22</span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># query(eh, "nullrangesdata")</span>
<span class="va">seg_cbs</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7307"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2022-07-22</span></code></pre>
<pre><code><span class="co">## see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seg_hmm</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7308"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span>
<span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seg</span> <span class="op">&lt;-</span> <span class="va">seg_cbs</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="segmentation-by-gene-density">Segmentation by gene density<a class="anchor" aria-label="anchor" href="#segmentation-by-gene-density"></a>
</h2>
<p>First we obtain the Ensembl genes <span class="citation">(Howe et al.
2020)</span> for segmenting by gene density. We obtain these using the
<em>ensembldb</em> package <span class="citation">(Rainer, Gatto, and
Weichenberger 2019)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jorainer/ensembldb" class="external-link">ensembldb</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span><span class="op">)</span>
<span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span>
<span class="va">filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilterList.html" class="external-link">AnnotationFilterList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationFilter/man/AnnotationFilter.html" class="external-link">GeneIdFilter</a></span><span class="op">(</span><span class="st">"ENSG"</span>, <span class="st">"startsWith"</span><span class="op">)</span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ensembldb/man/EnsDb-exonsBy.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="va">filt</span><span class="op">)</span></code></pre></div>
<p>We perform some processing to align the sequences (chromosomes) of
<code>g</code> with our excluded regions and our features of interest
(DNase hypersensitive sites, or DHS, defined below).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, <span class="st">"MT"</span><span class="op">)</span>
<span class="co"># normally we would assign a new style, but for recent host issues</span>
<span class="co">## seqlevelsStyle(g) &lt;- "UCSC" </span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg38"</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">sortSeqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span>
<span class="co">##  5194  3971  3010  2505  2868  2863  2867  2353  2242  2204  3235  2940  1304 </span>
<span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span>
<span class="co">##  2224  2152  2511  2995  1170  2926  1386   835  1318  2359   523</span></code></pre>
<div class="section level3">
<h3 id="import-excluded-regions">Import excluded regions<a class="anchor" aria-label="anchor" href="#import-excluded-regions"></a>
</h3>
<p>We next import excluded regions including ENCODE-produced excludable
regions<span class="citation">(Amemiya, Kundaje, and Boyle 2019)</span>,
telomeres from UCSC, centromeres <span class="citation">(Commo
2022)</span>. For easy use, pre-combined excludable regions is stored in
<em>ExperimentHub.</em> These steps using <em>excluderanges</em> package
<span class="citation">(Dozmorov et al. 2022)</span> are included in
<em>nullrangesData</em> in the
<code>inst/scripts/make-segmentation-hg38.R</code> script.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># suppressPackageStartupMessages(library(AnnotationHub))</span>
<span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7306"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] TRUE</span></code></pre>
</div>
<div class="section level3">
<h3 id="cbs-segmentation">CBS segmentation<a class="anchor" aria-label="anchor" href="#cbs-segmentation"></a>
</h3>
<p>We first demonstrate the use a CBS segmentation as implemented in
<em>DNAcopy</em> <span class="citation">(Olshen et al. 2004)</span>.</p>
<p>We load the <em>nullranges</em> and <em>plyranges</em> packages, and
<em>patchwork</em> in order to produce grids of plots.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges">nullranges</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></code></pre></div>
<p>We subset the excluded ranges to those which are 500 bp or larger.
The motivation for this step is to avoid segmenting the genome into many
small pieces due to an abundance of short excluded regions. Note that we
re-save the excluded ranges to <code>exclude</code>.</p>
<p>Here, and below, we need to specify <code><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">plyranges::filter</a></code> as
it conflicts with <code>filter</code> exported by
<em>ensembldb</em>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>
<span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">exclude</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span>
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">500</span><span class="op">)</span>
<span class="va">L_s</span> <span class="op">&lt;-</span> <span class="fl">1e6</span>
<span class="va">seg_cbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,
                          exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"cbs"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Analyzing: Sample.1</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_cbs</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-cbs-1.png"><!-- --></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-cbs-2.png"><!-- --></p>
<p>Note here, the default <em>ranges</em> plot gives whole genome and
every fractured bind regions represents state transformations happens.
However, some transformations within small ranges cannot be visualized,
e.g 1kb. If user want to look into specific ranges of segmentation
state, the <em>region</em> argument is flexible to support.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr16"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">3e7</span>,<span class="fl">4e7</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_cbs</span>, <span class="va">exclude</span>, type<span class="op">=</span><span class="st">"ranges"</span>, region<span class="op">=</span><span class="va">region</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/unnamed-chunk-7-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="alternatively-hmm-segmentation">Alternatively: HMM segmentation<a class="anchor" aria-label="anchor" href="#alternatively-hmm-segmentation"></a>
</h3>
<p>Here we use an alternative segmentation implemented in the
<em>RcppHMM</em> CRAN package, using the <code>initGHMM</code>,
<code>learnEM</code>, and <code>viterbi</code> functions.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seg_hmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,
                          exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"hmm"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Finished at Iteration: 111 with Error: 9.22278e-06</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_hmm</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-hmm-1.png"><!-- --></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-hmm-2.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="segmented-block-bootstrap">Segmented block bootstrap<a class="anchor" aria-label="anchor" href="#segmented-block-bootstrap"></a>
</h2>
<p>We use a set of DNase hypersensitivity sites (DHS) from the ENCODE
project <span class="citation">(ENCODE 2012)</span> in A549 cell line
(ENCSR614GWM). Here, for speed, we work with a pre-processed data object
from ExperimentHub, created using the following steps:</p>
<ul>
<li>Download ENCODE DNase hypersensitive peaks in A549 from
<em>AnnotationHub</em>
</li>
<li>Subset to standard chromosomes and remove mitochondrial DNA</li>
<li>Use a chain file from UCSC to lift ranges from hg19 to hg38</li>
<li>Sort the DHS features to be bootstrapped</li>
</ul>
<p>These steps are included in <em>nullrangesData</em> in the
<code>inst/scripts/make-dhs-data.R</code> script.</p>
<p>For speed of the vignette, we restrict to a smaller number of DHS,
filtering by the signal value. We also remove metadata columns that we
don’t need for the bootstrap analysis. Consider, when creating
bootstrapped data, that you will be creating an object many times larger
than your original features, so filtering and trimming extra metadata
can help make the analysis more efficient.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span><span class="op">)</span>
<span class="va">dhs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/DHSA549Hg38.html" class="external-link">DHSA549Hg38</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signalValue</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span>
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">select</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 6214</span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span>
<span class="co">##  1436   252   108    30   148    51   184   146   155   443   436   526    20 </span>
<span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span>
<span class="co">##   197   265   214   715    20   649   142    31    19    17    10</span></code></pre>
<p>Now we apply a segmented block bootstrap with blocks of size 500kb,
to the peaks. Here we show generation of 50 iterations of a block
bootstrap followed by a typical overlap analysis using
<em>plyranges</em> <span class="citation">(Lee, Cook, and Lawrence
2019)</span>. (We might normally do 100 iterations or more, depending on
the granularity of the bootstrap distribution that is needed.)</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co"># for reproducibility</span>
<span class="va">R</span> <span class="op">&lt;-</span> <span class="fl">50</span>
<span class="va">blockLength</span> <span class="op">&lt;-</span> <span class="fl">5e5</span>
<span class="va">boots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, R <span class="op">=</span> <span class="va">R</span>, seg <span class="op">=</span> <span class="va">seg</span>, exclude<span class="op">=</span><span class="va">exclude</span><span class="op">)</span>
<span class="va">boots</span></code></pre></div>
<pre><code><span class="co">## bootRanges object with 310726 ranges and 4 metadata columns:</span>
<span class="co">##            seqnames            ranges strand |        id     block  iter</span>
<span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;Rle&gt;</span>
<span class="co">##        [1]     chr1     242791-242940      * |       347         5     1</span>
<span class="co">##        [2]     chr1     256031-256180      * |       348         5     1</span>
<span class="co">##        [3]     chr1     391535-391684      * |      5301         8     1</span>
<span class="co">##        [4]     chr1     421046-421195      * |      5302         8     1</span>
<span class="co">##        [5]     chr1     438186-438335      * |      5303         8     1</span>
<span class="co">##        ...      ...               ...    ... .       ...       ...   ...</span>
<span class="co">##   [310722]     chrY 27090908-27091057      * |      2133     12441    50</span>
<span class="co">##   [310723]     chrY 27194968-27195117      * |      2134     12441    50</span>
<span class="co">##   [310724]     chrY 27224188-27224337      * |      2135     12441    50</span>
<span class="co">##   [310725]     chrY 27234153-27234302      * |      2136     12441    50</span>
<span class="co">##   [310726]     chrY 27789879-27790028      * |      2116     12442    50</span>
<span class="co">##            blockLength</span>
<span class="co">##                  &lt;Rle&gt;</span>
<span class="co">##        [1]      500000</span>
<span class="co">##        [2]      500000</span>
<span class="co">##        [3]      500000</span>
<span class="co">##        [4]      500000</span>
<span class="co">##        [5]      500000</span>
<span class="co">##        ...         ...</span>
<span class="co">##   [310722]      500000</span>
<span class="co">##   [310723]      500000</span>
<span class="co">##   [310724]      500000</span>
<span class="co">##   [310725]      500000</span>
<span class="co">##   [310726]      500000</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 24 sequences from hg38 genome</span></code></pre>
<p>What is returned here? The <code>bootRanges</code> function returns a
<em>bootRanges</em> object, which is a simple sub-class of
<em>GRanges</em>. The iteration (<code>iter</code>) and the block length
(<code>blockLength</code>) are recorded as metadata columns, accessible
via <code>mcols</code>. We return the bootstrapped ranges as
<em>GRanges</em> rather than <em>GRangesList</em>, as the former is more
compatible with downstream overlap joins with <em>plyranges</em>, where
the iteration column can be used with <code>group_by</code> to provide
per bootstrap summary statistics, as shown below.</p>
<p>Note that we use the <code>exclude</code> object from the previous
step, which does not contain small ranges. If one wanted to also avoid
generation of bootstrapped features that overlap small excluded ranges,
then omit this filtering step (use the original, complete
<code>exclude</code> feature set).</p>
<p>We can examine the number and extent of the bootstrapped data:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">boots</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">summarize</a></span><span class="op">(</span>
    n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>,
    total_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">width</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 50 rows and 3 columns</span>
<span class="co">##      iter         n total_width</span>
<span class="co">##     &lt;Rle&gt; &lt;integer&gt;   &lt;integer&gt;</span>
<span class="co">## 1       1      6150      923802</span>
<span class="co">## 2       2      6276      945143</span>
<span class="co">## 3       3      6249      939485</span>
<span class="co">## 4       4      6293      946714</span>
<span class="co">## 5       5      6365      957108</span>
<span class="co">## ...   ...       ...         ...</span>
<span class="co">## 46     46      6214      936107</span>
<span class="co">## 47     47      6177      930282</span>
<span class="co">## 48     48      6225      936523</span>
<span class="co">## 49     49      6185      931428</span>
<span class="co">## 50     50      6250      937598</span></code></pre>
<p>which is similar to the orig inal set, after filtering out those
overlapping the excluded region:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dhs_filt</span> <span class="op">&lt;-</span> <span class="va">dhs</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-filter-overlaps.html" class="external-link">filter_by_non_overlaps</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span>
<span class="va">dhs_filt</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, total_width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">width</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 1 row and 2 columns</span>
<span class="co">##           n total_width</span>
<span class="co">##   &lt;integer&gt;   &lt;integer&gt;</span>
<span class="co">## 1      6166      927017</span></code></pre>
</div>
<div class="section level2">
<h2 id="use-with-plyranges">Use with <em>plyranges</em><a class="anchor" aria-label="anchor" href="#use-with-plyranges"></a>
</h2>
<p>Suppose we have a set of features <code>x</code> and we are
interested in evaluating the overlap of this set with the DHS. We can
calculate for example the mean observed number of overlaps for features
in <code>x</code> (or something more complicated, e.g. the maximum log
fold change or signal value for DHS peaks within a <code>maxgap</code>
window of <code>x</code>).</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fl">50</span><span class="op">:</span><span class="fl">99</span> <span class="op">*</span> <span class="fl">1e6</span>, width<span class="op">=</span><span class="fl">1e6</span><span class="op">)</span>, x_id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">mutate</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-count-overlaps.html" class="external-link">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">dhs_filt</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span> <span class="va">x</span><span class="op">$</span><span class="va">n_overlaps</span> <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 1.26</span></code></pre>
<p>We can repeat this with the bootstrapped features using a
<code>group_by</code> command, a <code>summarize</code>, followed by a
second <code>group_by</code> and <code>summarize</code>. It may help to
step through these commands one by one to understand what the
intermediate output is.</p>
<p>Note that we need to use <code><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">tidyr::complete</a></code> in order to
fill in combinations of <code>x</code> and <code>iter</code> where the
overlap was 0.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">boot_stats</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_inner</a></span><span class="op">(</span><span class="va">boots</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">summarize</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/n.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html" class="external-link">complete</a></span><span class="op">(</span><span class="va">x_id</span>, <span class="va">iter</span>, fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">iter</span><span class="op">)</span> <span class="op"><a href="https://tidyr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">summarize</a></span><span class="op">(</span>meanOverlaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">n_overlaps</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The above code, first grouping by <code>x_id</code> and
<code>iter</code>, then subsequently by <code>iter</code> is general and
allows for more complex analysis than just mean overlap (e.g. how many
times an <code>x</code> range has 1 or more overlap, what is the mean or
max signal value for peaks overlapping ranges in <code>x</code>,
etc.).</p>
<p>Finally we can plot a histogram. In this case, as the <code>x</code>
features were arbitrary, our observed value falls within the
distribution of mean overlap with bootstrapped data.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">boot_stats</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">meanOverlaps</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth<span class="op">=</span><span class="fl">.2</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/boot-hist-1.png"><!-- --></p>
<p>For more examples of combining <code>bootRanges</code> from
<em>nullranges</em> with <em>plyranges</em> piped operations, see the
relevant chapter in the <a href="https://nullranges.github.io/tidy-ranges-tutorial/bootstrap-overlap.html" class="external-link">tidy-ranges-tutorial</a>
book.</p>
</div>
<div class="section level2">
<h2 id="timing-on-dhs-peaks">Timing on DHS peaks<a class="anchor" aria-label="anchor" href="#timing-on-dhs-peaks"></a>
</h2>
<p>Here, we test the speed of the various options for bootstrapping (see
below for visualization of the difference).</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshuaulrich/microbenchmark/" class="external-link">microbenchmark</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html" class="external-link">microbenchmark</a></span><span class="op">(</span>
  list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">alist</a></span><span class="op">(</span>
    prop <span class="op">=</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, seg <span class="op">=</span> <span class="va">seg</span>, proportionLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
    no_prop <span class="op">=</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, seg <span class="op">=</span> <span class="va">seg</span>, proportionLength <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">)</span>, times<span class="op">=</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Unit: milliseconds</span>
<span class="co">##     expr     min      lq     mean   median      uq      max neval</span>
<span class="co">##     prop 87.5933 90.7356 94.24532 91.87500 92.3701 119.8743    10</span>
<span class="co">##  no_prop 80.9131 81.2898 82.67386 82.05245 83.8276  85.6265    10</span></code></pre>
</div>
<div class="section level2">
<h2 id="visualizing-toy-bootstrap-samples">Visualizing toy bootstrap samples<a class="anchor" aria-label="anchor" href="#visualizing-toy-bootstrap-samples"></a>
</h2>
<p>Below we present a toy example for visualizing the segmented block
bootstrap. First, we define a helper function for plotting
<em>GRanges</em> using <em>plotgardener</em> <span class="citation">(Kramer et al. 2021)</span>. A key aspect here is that
we color the original and bootstrapped ranges by the genomic state (the
state of the segmentation that the original ranges fall in).</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener" class="external-link">plotgardener</a></span><span class="op">)</span><span class="op">)</span>
<span class="va">my_palette</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"green3"</span>,<span class="st">"red3"</span>,<span class="st">"dodgerblue"</span>,
         <span class="st">"blue2"</span>,<span class="st">"green4"</span>,<span class="st">"darkred"</span><span class="op">)</span>, <span class="va">n</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">plotGRanges</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html" class="external-link">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">5</span>, xgrid <span class="op">=</span> <span class="fl">0</span>,
                ygrid <span class="op">=</span> <span class="fl">0</span>, showGuides <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
    <span class="va">chromend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[[</span><span class="va">chrom</span><span class="op">]</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span>
      <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pgParams.html" class="external-link">pgParams</a></span><span class="op">(</span>chromstart <span class="op">=</span> <span class="fl">0</span>, chromend <span class="op">=</span> <span class="va">chromend</span>,
                    x <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="va">chromend</span><span class="op">/</span><span class="fl">500</span>, height <span class="op">=</span> <span class="fl">2</span>,
                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">chromend</span>, <span class="fl">50</span><span class="op">)</span>,
                    fill <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/colorby.html" class="external-link">colorby</a></span><span class="op">(</span><span class="st">"state_col"</span>, palette<span class="op">=</span><span class="va">my_palette</span><span class="op">)</span><span class="op">)</span>
      <span class="va">prngs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotRanges.html" class="external-link">plotRanges</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gr</span>, params <span class="op">=</span> <span class="va">p</span>,
                          chrom <span class="op">=</span> <span class="va">chrom</span>,
                          y <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span>,
                          just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span><span class="op">)</span>
      <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoGenomeLabel.html" class="external-link">annoGenomeLabel</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">prngs</span>, params <span class="op">=</span> <span class="va">p</span>, y <span class="op">=</span> <span class="fl">0.1</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span></code></pre></div>
<p>Create a toy genome segmentation:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span>
<span class="va">seq_nms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="va">seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>
  seqnames <span class="op">=</span> <span class="va">seq_nms</span>,
  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">101</span>, <span class="fl">201</span>, <span class="fl">401</span>, <span class="fl">1</span>, <span class="fl">201</span>, <span class="fl">301</span><span class="op">)</span>,
          width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>,
  seqlengths<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>chr1<span class="op">=</span><span class="fl">500</span>,chr2<span class="op">=</span><span class="fl">400</span><span class="op">)</span>,
  state <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>,
  state_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>We can visualize with our helper function:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">seg</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toysegments-1.png"><!-- --></p>
<p>Now create small ranges distributed uniformly across the toy
genome:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">200</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>
  seqnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="va">n</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">500</span><span class="op">-</span><span class="fl">10</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr2"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">400</span><span class="op">)</span><span class="op">]</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">seg</span>, type<span class="op">=</span><span class="st">"within"</span>, select<span class="op">=</span><span class="st">"first"</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">idx</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span>
<span class="va">gr</span><span class="op">$</span><span class="va">state</span> <span class="op">&lt;-</span> <span class="va">seg</span><span class="op">$</span><span class="va">state</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span>
<span class="va">gr</span><span class="op">$</span><span class="va">state_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seg</span><span class="op">$</span><span class="va">state_col</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toyranges-1.png"><!-- --></p>
<div class="section level3">
<h3 id="not-scaling-by-segmentation">Not scaling by segmentation<a class="anchor" aria-label="anchor" href="#not-scaling-by-segmentation"></a>
</h3>
<p>We can visualize block bootstrapped ranges when the blocks do not
scale to segment state length:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">25</span>, seg <span class="op">=</span> <span class="va">seg</span>,
                       proportionLength <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toy-no-prop-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="scaling-by-segmentation">Scaling by segmentation<a class="anchor" aria-label="anchor" href="#scaling-by-segmentation"></a>
</h3>
<p>This time the blocks scale to the segment state length. Note that in
this case <code>blockLength</code> is the <em>maximal</em> block length
possible, but the actual block lengths per segment will be smaller
(proportional to the fraction of basepairs covered by that state in the
genome segmentation).</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">50</span>, seg <span class="op">=</span> <span class="va">seg</span>,
                       proportionLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toy-prop-1.png"><!-- --></p>
<p>Note that some ranges from adjacent states are allowed to be placed
within different states in the bootstrap sample. This is because, during
the random sampling of blocks of original data, a block is allowed to
extend beyond the segmentation region of the state being sampled, and
features from adjacent states are not excluded from the sampled
block.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.2.0 (2022-04-22)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] plotgardener_1.3.8          microbenchmark_1.4.9       </span>
<span class="co">##  [3] ggplot2_3.3.6               tidyr_1.2.0                </span>
<span class="co">##  [5] patchwork_1.1.1             plyranges_1.17.0           </span>
<span class="co">##  [7] nullranges_1.3.4            EnsDb.Hsapiens.v86_2.99.0  </span>
<span class="co">##  [9] ensembldb_2.21.3            AnnotationFilter_1.21.0    </span>
<span class="co">## [11] GenomicFeatures_1.49.5      AnnotationDbi_1.59.1       </span>
<span class="co">## [13] nullrangesData_1.3.0        InteractionSet_1.25.0      </span>
<span class="co">## [15] SummarizedExperiment_1.27.1 Biobase_2.57.1             </span>
<span class="co">## [17] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span>
<span class="co">## [19] GenomicRanges_1.49.0        GenomeInfoDb_1.33.3        </span>
<span class="co">## [21] IRanges_2.31.0              S4Vectors_0.35.1           </span>
<span class="co">## [23] ExperimentHub_2.5.0         AnnotationHub_3.5.0        </span>
<span class="co">## [25] BiocFileCache_2.5.0         dbplyr_2.2.1               </span>
<span class="co">## [27] BiocGenerics_0.43.1        </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.4             plyr_1.8.7                   </span>
<span class="co">##   [3] RcppHMM_1.2.2                 lazyeval_0.2.2               </span>
<span class="co">##   [5] BiocParallel_1.31.12          digest_0.6.29                </span>
<span class="co">##   [7] yulab.utils_0.0.5             htmltools_0.5.2              </span>
<span class="co">##   [9] fansi_1.0.3                   magrittr_2.0.3               </span>
<span class="co">##  [11] memoise_2.0.1                 ks_1.13.5                    </span>
<span class="co">##  [13] Biostrings_2.65.1             pkgdown_2.0.4                </span>
<span class="co">##  [15] prettyunits_1.1.1             colorspace_2.0-3             </span>
<span class="co">##  [17] blob_1.2.3                    rappdirs_0.3.3               </span>
<span class="co">##  [19] textshaping_0.3.6             xfun_0.31                    </span>
<span class="co">##  [21] dplyr_1.0.9                   crayon_1.5.1                 </span>
<span class="co">##  [23] RCurl_1.98-1.7                jsonlite_1.8.0               </span>
<span class="co">##  [25] glue_1.6.2                    gtable_0.3.0                 </span>
<span class="co">##  [27] zlibbioc_1.43.0               XVector_0.37.0               </span>
<span class="co">##  [29] strawr_0.0.9                  DelayedArray_0.23.1          </span>
<span class="co">##  [31] scales_1.2.0                  mvtnorm_1.1-3                </span>
<span class="co">##  [33] DBI_1.1.3                     Rcpp_1.0.8.3                 </span>
<span class="co">##  [35] xtable_1.8-4                  progress_1.2.2               </span>
<span class="co">##  [37] gridGraphics_0.5-1            bit_4.0.4                    </span>
<span class="co">##  [39] mclust_5.4.10                 httr_1.4.3                   </span>
<span class="co">##  [41] RColorBrewer_1.1-3            speedglm_0.3-4               </span>
<span class="co">##  [43] ellipsis_0.3.2                pkgconfig_2.0.3              </span>
<span class="co">##  [45] XML_3.99-0.10                 farver_2.1.0                 </span>
<span class="co">##  [47] sass_0.4.1                    utf8_1.2.2                   </span>
<span class="co">##  [49] DNAcopy_1.71.0                ggplotify_0.1.0              </span>
<span class="co">##  [51] tidyselect_1.1.2              labeling_0.4.2               </span>
<span class="co">##  [53] rlang_1.0.3                   later_1.3.0                  </span>
<span class="co">##  [55] munsell_0.5.0                 BiocVersion_3.16.0           </span>
<span class="co">##  [57] tools_4.2.0                   cachem_1.0.6                 </span>
<span class="co">##  [59] cli_3.3.0                     generics_0.1.2               </span>
<span class="co">##  [61] RSQLite_2.2.14                ggridges_0.5.3               </span>
<span class="co">##  [63] evaluate_0.15                 stringr_1.4.0                </span>
<span class="co">##  [65] fastmap_1.1.0                 yaml_2.3.5                   </span>
<span class="co">##  [67] ragg_1.2.2                    knitr_1.39                   </span>
<span class="co">##  [69] bit64_4.0.5                   fs_1.5.2                     </span>
<span class="co">##  [71] purrr_0.3.4                   KEGGREST_1.37.3              </span>
<span class="co">##  [73] mime_0.12                     pracma_2.3.8                 </span>
<span class="co">##  [75] xml2_1.3.3                    biomaRt_2.53.2               </span>
<span class="co">##  [77] compiler_4.2.0                filelock_1.0.2               </span>
<span class="co">##  [79] curl_4.3.2                    png_0.1-7                    </span>
<span class="co">##  [81] interactiveDisplayBase_1.35.0 tibble_3.1.7                 </span>
<span class="co">##  [83] bslib_0.3.1                   stringi_1.7.6                </span>
<span class="co">##  [85] highr_0.9                     desc_1.4.1                   </span>
<span class="co">##  [87] lattice_0.20-45               ProtGenerics_1.29.0          </span>
<span class="co">##  [89] Matrix_1.4-1                  vctrs_0.4.1                  </span>
<span class="co">##  [91] pillar_1.7.0                  lifecycle_1.0.1              </span>
<span class="co">##  [93] BiocManager_1.30.18           jquerylib_0.1.4              </span>
<span class="co">##  [95] data.table_1.14.2             bitops_1.0-7                 </span>
<span class="co">##  [97] httpuv_1.6.5                  rtracklayer_1.57.0           </span>
<span class="co">##  [99] R6_2.5.1                      BiocIO_1.7.1                 </span>
<span class="co">## [101] promises_1.2.0.1              KernSmooth_2.23-20           </span>
<span class="co">## [103] codetools_0.2-18              MASS_7.3-57                  </span>
<span class="co">## [105] assertthat_0.2.1              rprojroot_2.0.3              </span>
<span class="co">## [107] rjson_0.2.21                  withr_2.5.0                  </span>
<span class="co">## [109] GenomicAlignments_1.33.1      Rsamtools_2.13.3             </span>
<span class="co">## [111] GenomeInfoDbData_1.2.8        parallel_4.2.0               </span>
<span class="co">## [113] hms_1.1.1                     grid_4.2.0                   </span>
<span class="co">## [115] rmarkdown_2.14                shiny_1.7.1                  </span>
<span class="co">## [117] restfulr_0.0.15</span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode_exclude" class="csl-entry">
Amemiya, Haley M, Anshul Kundaje, and Alan P Boyle. 2019. <span>“The
<span>ENCODE</span> Blacklist: Identification of Problematic Regions of
the Genome.”</span> <em>Scientific Reports</em> 9 (1): 9354. <a href="https://doi.org/10.1038/s41598-019-45839-z" class="external-link">https://doi.org/10.1038/s41598-019-45839-z</a>.
</div>
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy
R. Zhang. 2010. <span>“Subsampling Methods for Genomic
Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4):
1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-rCGH" class="csl-entry">
Commo, Frederic. 2022. <em>rCGH: Comprehensive Pipeline for Analyzing
and Visualizing Array-Based CGH Data</em>. <a href="https://github.com/fredcommo/rCGH" class="external-link">https://github.com/fredcommo/rCGH</a>.
</div>
<div id="ref-excluderanges" class="csl-entry">
Dozmorov, Mikhail G., Eric Davis, Wancen Mu, Stuart Lee, Tim Triche,
Douglas Phanstiel, and Michael Love. 2022. <em>Excluderanges</em>. <a href="https://github.com/mdozmorov/excluderanges" class="external-link">https://github.com/mdozmorov/excluderanges</a>.
</div>
<div id="ref-encode" class="csl-entry">
ENCODE. 2012. <span>“<span class="nocase">An integrated encyclopedia of
DNA elements in the human genome</span>.”</span> <em>Nature</em> 489
(7414): 57–74. <a href="https://doi.org/10.1038/nature11247" class="external-link">https://doi.org/10.1038/nature11247</a>.
</div>
<div id="ref-ensembl2021" class="csl-entry">
Howe, Kevin L, Premanand Achuthan, James Allen, Jamie Allen, Jorge
Alvarez-Jarreta, M Ridwan Amode, Irina M Armean, et al. 2020.
<span>“<span>Ensembl 2021</span>.”</span> <em>Nucleic Acids
Research</em> 49 (D1): D884–91. <a href="https://doi.org/10.1093/nar/gkaa942" class="external-link">https://doi.org/10.1093/nar/gkaa942</a>.
</div>
<div id="ref-Kramer2021" class="csl-entry">
Kramer, Nicole E, Eric S Davis, Craig D Wenger, Erika M Deoudes, Sarah M
Parker, Michael I Love, and Douglas H Phanstiel. 2021. <span>“<span class="nocase">Plotgardener: Cultivating precise multi-panel figures in
R</span>.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.09.08.459338" class="external-link">https://doi.org/10.1101/2021.09.08.459338</a>.
</div>
<div id="ref-Lee2019" class="csl-entry">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges:
A Grammar of Genomic Data Transformation.”</span> <em>Genome
Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8" class="external-link">https://doi.org/10.1186/s13059-018-1597-8</a>.
</div>
<div id="ref-dnacopy" class="csl-entry">
Olshen, A. B., E. S. Venkatraman, R. Lucito, and M. Wigler. 2004.
<span>“<span class="nocase"><span>C</span>ircular binary segmentation
for the analysis of array-based
<span>D</span><span>N</span><span>A</span> copy number
data</span>.”</span> <em>Biostatistics</em> 5 (4): 557–72.
</div>
<div id="ref-ensembldb" class="csl-entry">
Rainer, Johannes, Laurent Gatto, and Christian X Weichenberger. 2019.
<span>“<span class="nocase">ensembldb: an R package to create and use
Ensembl-based annotation resources</span>.”</span>
<em>Bioinformatics</em> 35 (17): 3151–53. <a href="https://doi.org/10.1093/bioinformatics/btz031" class="external-link">https://doi.org/10.1093/bioinformatics/btz031</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://twitter.com/wancenm" class="external-link">Wancen Mu</a>, <a href="https://ericscottdavis.com" class="external-link">Eric Davis</a>, <a href="http://phanstiel-lab.med.unc.edu" class="external-link">Douglas Phanstiel</a>, <a href="https://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://chanzuckerberg.com" class="external-link">CZI</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
