<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Segmented block bootstrap • nullranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Segmented block bootstrap">
<meta property="og:description" content="nullranges">
<meta property="og:image" content="https://nullranges.github.io/nullranges/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nullranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/nullranges.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/matching_ranges.html">Overview of matchRanges</a>
    </li>
    <li>
      <a href="../articles/matching_granges.html">Case study I: CTCF occupancy</a>
    </li>
    <li>
      <a href="../articles/matching_ginteractions.html">Case study II: CTCF orientation</a>
    </li>
    <li>
      <a href="../articles/segmented_boot_ranges.html">Segmented block bootstrap</a>
    </li>
    <li>
      <a href="../articles/unseg_boot_ranges.html">Unsegmented block bootstrap</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Segmented block bootstrap</h1>
                        <h4 data-toc-skip class="author">Wancen Mu</h4>
            
            <h4 data-toc-skip class="date">12/14/2021</h4>
      
      
      <div class="hidden name"><code>segmented_boot_ranges.Rmd</code></div>

    </div>

    
    
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>The following vignette describes the <em>nullranges</em> implementation of the block bootstrap with respect to a genomic segmentation. See the main nullranges vignette for an overview of the idea of bootstrapping, and there is additionally a vignette on block boostrapping without respect to segmentation.</p>
<p>As proposed by <span class="citation">Bickel et al. (2010)</span>, <em>nullranges</em> contains an implementation of a block bootstrap, such that features are sampled from the genome in blocks. Blocks are sampled and placed within regions of a genome <em>segmentation</em>. That is, for a genome segmented into states 1,2,…,S, blocks from state <em>s</em> will be used to tile the ranges of state <em>s</em> in each bootstrap sample.</p>
<p>The segmented block bootstrap has two options, either:</p>
<ul>
<li>Perform a de-novo segmentation of the genome using feature density, e.g. gene density</li>
<li>Use exiting segmentation (e.g. ChromHMM, etc.) downloaded from AnnotationHub or external to Bioconductor (BED files imported with <em>rtracklayer</em>)</li>
</ul>
<p>In this vignette, we give an example of segmenting the hg38 genome by Ensembl gene density, then we profile the time to generate a single block bootstrap sample. Finally, we use a toy dataset to visualize what a segmented block bootstrap sample looks like with respect to a genome segmentation. Future versions of this vignette will demonstrate the functions used within an overlap analysis. See also the unsegmented block bootstrap vignette in <em>nullranges</em>, if it is not desired to bootstrap with respect to a genome segmentation.</p>
<p>A finally consideration is whether the blocks should scale proportionally to the segment state length, with the default setting of <code>proportionLength=TRUE</code>. When blocks scale proportionally, <code>blockLength</code> provides the maximal length of a block, while the actual block length used for a segmentation state is proportional to the fraction of genomic basepairs covered by that state. This option is visualized on toy data at the end of this vignette.</p>
</div>
<div class="section level1">
<h1 id="segmentation-by-gene-density">Segmentation by gene density<a class="anchor" aria-label="anchor" href="#segmentation-by-gene-density"></a>
</h1>
<p>First we obtain the Ensembl genes <span class="citation">(Howe et al. 2020)</span> for segmenting by gene density. We obtain these using the <em>ensembldb</em> package <span class="citation">(Rainer, Gatto, and Weichenberger 2019)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span><span class="op">)</span>
<span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span>
<span class="va">filt</span> <span class="op">&lt;-</span> <span class="fu">AnnotationFilterList</span><span class="op">(</span><span class="fu">GeneIdFilter</span><span class="op">(</span><span class="st">"ENSG"</span>, <span class="st">"startsWith"</span><span class="op">)</span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ensembldb/man/EnsDb-exonsBy.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">edb</span>, filter <span class="op">=</span> <span class="va">filt</span><span class="op">)</span></code></pre></div>
<p>We perform some processing to align the sequences (chromosomes) of <code>g</code> with our excluded regions and our features of interest (DNase hypersensitive sites, or DHS, defined below).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomeInfoDb" class="external-link">GenomeInfoDb</a></span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, <span class="st">"MT"</span><span class="op">)</span>
<span class="co"># normally we would assign a new style, but for recent host issues</span>
<span class="co">## seqlevelsStyle(g) &lt;- "UCSC" </span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"hg38"</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">sortSeqlevels</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span>
<span class="co">##  5194  3971  3010  2505  2868  2863  2867  2353  2242  2204  3235  2940  1304 </span>
<span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span>
<span class="co">##  2224  2152  2511  2995  1170  2926  1386   835  1318  2359   523</span></code></pre>
<div class="section level2">
<h2 id="import-excluded-regions">Import excluded regions<a class="anchor" aria-label="anchor" href="#import-excluded-regions"></a>
</h2>
<p>We next import excluded regions, as created by <span class="citation">Amemiya, Kundaje, and Boyle (2019)</span>, and packaged for R/Bioconductor in the <em>excluderanges</em> package by Mikhail Dozmorov.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mdozmorov/excluderanges" class="external-link">excluderanges</a></span><span class="op">)</span>
<span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-10-20</span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># `excludeGR.hg38.Kundaje.1` -- see excluderanges vignette</span>
<span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH95917"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] TRUE</span></code></pre>
</div>
<div class="section level2">
<h2 id="cbs-segmentation">CBS segmentation<a class="anchor" aria-label="anchor" href="#cbs-segmentation"></a>
</h2>
<p>We first demonstrate the use a CBS segmentation as implemented in <em>DNAcopy</em> <span class="citation">(Olshen et al. 2004)</span>.</p>
<p>We load the <em>nullranges</em> and <em>plyranges</em> packages, and <em>patchwork</em> in order to produce grids of plots.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nullranges.github.io/nullranges">nullranges</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></code></pre></div>
<p>We subset the excluded ranges to those which are 500 bp or larger. The motivation for this step is to avoid segmenting the genome into many small pieces due to an abundance of short excluded regions. Note that we re-save the excluded ranges to <code>exclude</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>
<span class="va">exclude</span> <span class="op">&lt;-</span> <span class="va">exclude</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span>
  <span class="fu">plyranges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">exclude</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">500</span><span class="op">)</span>
<span class="va">L_s</span> <span class="op">&lt;-</span> <span class="fl">1e6</span>
<span class="va">seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,
                      exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"cbs"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Analyzing: Sample.1</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-cbs-1.png"><!-- --></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-cbs-2.png"><!-- --></p>
<p>Note here, the default <em>ranges</em> plot gives whole genome and every fractured bind regions represents state transformations happens. However, some transformations within small ranges cannot be visualized, e.g 1kb. If user want to look into specific ranges of segmentation state, the <em>region</em> argument is flexible to support.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr16"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">3e7</span>,<span class="fl">4e7</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg</span>, <span class="va">exclude</span>, type<span class="op">=</span><span class="st">"ranges"</span>, region<span class="op">=</span><span class="va">region</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/unnamed-chunk-5-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="alternatively-hmm-segmentation">Alternatively: HMM segmentation<a class="anchor" aria-label="anchor" href="#alternatively-hmm-segmentation"></a>
</h2>
<p>Here we use an alternative segmentation implemented in the <em>RcppHMM</em> CRAN package, using the <code>initGHMM</code>, <code>learnEM</code>, and <code>viterbi</code> functions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seg_hmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segmentDensity.html">segmentDensity</a></span><span class="op">(</span><span class="va">g</span>, n <span class="op">=</span> <span class="fl">3</span>, L_s <span class="op">=</span> <span class="va">L_s</span>,
                          exclude <span class="op">=</span> <span class="va">exclude</span>, type <span class="op">=</span> <span class="st">"hmm"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Finished at Iteration: 100 with Error: 9.31905e-06</span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ranges"</span>,<span class="st">"barplot"</span>,<span class="st">"boxplot"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/plotSegment.html">plotSegment</a></span><span class="op">(</span><span class="va">seg_hmm</span>, <span class="va">exclude</span>, type <span class="op">=</span> <span class="va">t</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-hmm-1.png"><!-- --></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/seg-hmm-2.png"><!-- --></p>
</div>
</div>
<div class="section level1">
<h1 id="segmented-block-bootstrap">Segmented block bootstrap<a class="anchor" aria-label="anchor" href="#segmented-block-bootstrap"></a>
</h1>
<p>We use a set of DNase hypersensitivity sites (DHS) from the ENCODE project <span class="citation">(ENCODE 2012)</span> in A549 cell line (ENCSR614GWM). Here, for speed, we work with a pre-processed data object from ExperimentHub, created using the following steps:</p>
<ul>
<li>Download ENCODE DNase hypersensitive peaks in A549 from <em>AnnotationHub</em>
</li>
<li>Subset to standard chromosomes and remove mitochondrial DNA</li>
<li>Use a chain file from UCSC to lift ranges from hg19 to hg38</li>
<li>Sort the DHS features to be bootstrapped</li>
</ul>
<p>These steps are included in <em>nullrangesData</em> in the <code>inst/scripts/make-dhs-data.R</code> script.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">nullrangesData</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-10-19</span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dhs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/nullrangesData/man/DHSA549Hg38.html" class="external-link">DHSA549Hg38</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## see ?nullrangesData and browseVignettes('nullrangesData') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">dhs</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 chr13 </span>
<span class="co">## 15240 14436  9549  6034  9714  8043  9549  8463  8247  8924  8618 10597  2732 </span>
<span class="co">## chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22  chrX  chrY </span>
<span class="co">##  5785  6929  6596 11438  2643  8818  5986  1471  3010  4024   239</span></code></pre>
<p>Now we apply a segmented block bootstrap with blocks of size 500kb, to the peaks. Here we simply show generation of two iterations of a block bootstrap, while we plan in future sections of this vignette to demonstrate its use in a workflow including <em>plyranges</em> <span class="citation">(Lee, Cook, and Lawrence 2019)</span>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blockLength</span> <span class="op">&lt;-</span> <span class="fl">5e5</span>
<span class="va">br</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, R <span class="op">=</span> <span class="fl">2</span>, seg <span class="op">=</span> <span class="va">seg</span>, exclude<span class="op">=</span><span class="va">exclude</span><span class="op">)</span>
<span class="va">br</span></code></pre></div>
<pre><code><span class="co">## bootRanges object with 354357 ranges and 9 metadata columns:</span>
<span class="co">##            seqnames            ranges strand |        name     score</span>
<span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">##        [1]     chr1 55046962-55047111      * |        &lt;NA&gt;         0</span>
<span class="co">##        [2]     chr1 55066027-55066176      * |        &lt;NA&gt;         0</span>
<span class="co">##        [3]     chr1 55112487-55112636      * |        &lt;NA&gt;         0</span>
<span class="co">##        [4]     chr1 55179402-55179551      * |        &lt;NA&gt;         0</span>
<span class="co">##        [5]     chr1 55179862-55180011      * |        &lt;NA&gt;         0</span>
<span class="co">##        ...      ...               ...    ... .         ...       ...</span>
<span class="co">##   [354353]    chr22 50428971-50429120      * |        &lt;NA&gt;         0</span>
<span class="co">##   [354354]    chr22 50465467-50465616      * |        &lt;NA&gt;         0</span>
<span class="co">##   [354355]    chr22 50517029-50517178      * |        &lt;NA&gt;         0</span>
<span class="co">##   [354356]    chr22 50534629-50534778      * |        &lt;NA&gt;         0</span>
<span class="co">##   [354357]    chr22 50788902-50789051      * |        &lt;NA&gt;         0</span>
<span class="co">##            signalValue    pValue    qValue      peak     block  iter</span>
<span class="co">##              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt; &lt;Rle&gt;</span>
<span class="co">##        [1]           3        -1        -1        -1         1     1</span>
<span class="co">##        [2]          18        -1        -1        -1         1     1</span>
<span class="co">##        [3]          13        -1        -1        -1         1     1</span>
<span class="co">##        [4]         166        -1        -1        -1         1     1</span>
<span class="co">##        [5]          41        -1        -1        -1         1     1</span>
<span class="co">##        ...         ...       ...       ...       ...       ...   ...</span>
<span class="co">##   [354353]          77        -1        -1        -1     12466     2</span>
<span class="co">##   [354354]          18        -1        -1        -1     12467     2</span>
<span class="co">##   [354355]          24        -1        -1        -1     12468     2</span>
<span class="co">##   [354356]          14        -1        -1        -1     12468     2</span>
<span class="co">##   [354357]          14        -1        -1        -1     12471     2</span>
<span class="co">##            blockLength</span>
<span class="co">##                  &lt;Rle&gt;</span>
<span class="co">##        [1]      500000</span>
<span class="co">##        [2]      500000</span>
<span class="co">##        [3]      500000</span>
<span class="co">##        [4]      500000</span>
<span class="co">##        [5]      500000</span>
<span class="co">##        ...         ...</span>
<span class="co">##   [354353]      500000</span>
<span class="co">##   [354354]      500000</span>
<span class="co">##   [354355]      500000</span>
<span class="co">##   [354356]      500000</span>
<span class="co">##   [354357]      500000</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 24 sequences from hg38 genome</span></code></pre>
<p><strong>What is returned here?</strong> The <code>bootRanges</code> function returns a <em>bootRanges</em> object, which is a simple sub-class of <em>GRanges</em>. The iteration (<code>iter</code>) and the block length (<code>blockLength</code>) are recorded as metadata columns, accessible via <code>mcols</code>. We chose to return the bootstrapped ranges as <em>GRanges</em> rather than <em>GRangesList</em>, as the former is more compatible with downstream overlap joins with <em>plyranges</em>, where the iteration column can be used with <code>group_by</code> to provide per bootstrap summary statistics.</p>
<p>Note that we use the <code>exclude</code> object from the previous step, which does not contain small ranges. If one wanted to also avoid generation of bootstrapped features that overlap small excluded ranges, then omit this filtering step (use the original, complete <code>exclude</code> feature set).</p>
</div>
<div class="section level1">
<h1 id="timing-on-dhs-peaks">Timing on DHS peaks<a class="anchor" aria-label="anchor" href="#timing-on-dhs-peaks"></a>
</h1>
<p>Here, we test the speed of the various options for bootstrapping (see below for visualization of the difference).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshuaulrich/microbenchmark/" class="external-link">microbenchmark</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/microbenchmark/man/microbenchmark.html" class="external-link">microbenchmark</a></span><span class="op">(</span>
  list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">alist</a></span><span class="op">(</span>
    prop <span class="op">=</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, seg <span class="op">=</span> <span class="va">seg</span>, proportionLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
    no_prop <span class="op">=</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">dhs</span>, <span class="va">blockLength</span>, seg <span class="op">=</span> <span class="va">seg</span>, proportionLength <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="op">)</span>, times<span class="op">=</span><span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Unit: milliseconds</span>
<span class="co">##     expr      min       lq     mean   median       uq      max neval</span>
<span class="co">##     prop 212.4731 234.4645 239.2914 237.6617 251.8102 258.3819    10</span>
<span class="co">##  no_prop 224.1936 225.3317 233.5321 231.5539 238.0041 251.3726    10</span></code></pre>
</div>
<div class="section level1">
<h1 id="visualizing-toy-bootstrap-samples">Visualizing toy bootstrap samples<a class="anchor" aria-label="anchor" href="#visualizing-toy-bootstrap-samples"></a>
</h1>
<p>Below we present a toy example for visualizing the segmented block bootstrap. First, we define a helper function for plotting <em>GRanges</em> using <em>plotgardener</em> <span class="citation">(Kramer et al. 2021)</span>. A key aspect here is that we color the original and bootstrapped ranges by the genomic state (the state of the segmentation that the original ranges fall in).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener" class="external-link">plotgardener</a></span><span class="op">)</span><span class="op">)</span>
<span class="va">my_palette</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>,<span class="st">"green3"</span>,<span class="st">"red3"</span>,<span class="st">"dodgerblue"</span>,
         <span class="st">"blue2"</span>,<span class="st">"green4"</span>,<span class="st">"darkred"</span><span class="op">)</span>, <span class="va">n</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">plotGRanges</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pageCreate.html" class="external-link">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">5</span>, xgrid <span class="op">=</span> <span class="fl">0</span>,
                ygrid <span class="op">=</span> <span class="fl">0</span>, showGuides <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
    <span class="va">chromend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">[[</span><span class="va">chrom</span><span class="op">]</span><span class="op">]</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="op">{</span>
      <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/pgParams.html" class="external-link">pgParams</a></span><span class="op">(</span>chromstart <span class="op">=</span> <span class="fl">0</span>, chromend <span class="op">=</span> <span class="va">chromend</span>,
                    x <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">4</span><span class="op">*</span><span class="va">chromend</span><span class="op">/</span><span class="fl">500</span>, height <span class="op">=</span> <span class="fl">2</span>,
                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">chromend</span>, <span class="fl">50</span><span class="op">)</span>,
                    fill <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/colorby.html" class="external-link">colorby</a></span><span class="op">(</span><span class="st">"state_col"</span>, palette<span class="op">=</span><span class="va">my_palette</span><span class="op">)</span><span class="op">)</span>
      <span class="va">prngs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/plotRanges.html" class="external-link">plotRanges</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">gr</span>, params <span class="op">=</span> <span class="va">p</span>,
                          chrom <span class="op">=</span> <span class="va">chrom</span>,
                          y <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span>,
                          just <span class="op">=</span> <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"bottom"</span><span class="op">)</span><span class="op">)</span>
      <span class="fu"><a href="https://phanstiellab.github.io/plotgardener/reference/annoGenomeLabel.html" class="external-link">annoGenomeLabel</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">prngs</span>, params <span class="op">=</span> <span class="va">p</span>, y <span class="op">=</span> <span class="fl">0.1</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">i</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span></code></pre></div>
<p>Create a toy genome segmentation:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span>
<span class="va">seq_nms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="va">seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>
  seqnames <span class="op">=</span> <span class="va">seq_nms</span>,
  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">101</span>, <span class="fl">201</span>, <span class="fl">401</span>, <span class="fl">1</span>, <span class="fl">201</span>, <span class="fl">301</span><span class="op">)</span>,
          width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">100</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>,
  seqlengths<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>chr1<span class="op">=</span><span class="fl">500</span>,chr2<span class="op">=</span><span class="fl">400</span><span class="op">)</span>,
  state <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">3</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>,
  state_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>We can visualize with our helper function:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotGRanges</span><span class="op">(</span><span class="va">seg</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toysegments-1.png"><!-- --></p>
<p>Now create small ranges distributed uniformly across the toy genome:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">200</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>
  seqnames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>,<span class="st">"chr2"</span><span class="op">)</span>, <span class="va">n</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">500</span><span class="op">-</span><span class="fl">10</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">10</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr2"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">400</span><span class="op">)</span><span class="op">]</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">seg</span>, type<span class="op">=</span><span class="st">"within"</span>, select<span class="op">=</span><span class="st">"first"</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">idx</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">]</span>
<span class="va">gr</span><span class="op">$</span><span class="va">state</span> <span class="op">&lt;-</span> <span class="va">seg</span><span class="op">$</span><span class="va">state</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span>
<span class="va">gr</span><span class="op">$</span><span class="va">state_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seg</span><span class="op">$</span><span class="va">state_col</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toyranges-1.png"><!-- --></p>
<div class="section level2">
<h2 id="not-scaling-by-segmentation">Not scaling by segmentation<a class="anchor" aria-label="anchor" href="#not-scaling-by-segmentation"></a>
</h2>
<p>We can visualize block bootstrapped ranges when the blocks do not scale to segment state length:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">25</span>, seg <span class="op">=</span> <span class="va">seg</span>,
                       proportionLength <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toy-no-prop-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="scaling-by-segmentation">Scaling by segmentation<a class="anchor" aria-label="anchor" href="#scaling-by-segmentation"></a>
</h2>
<p>This time the blocks scale to the segment state length. Note that in this case <code>blockLength</code> is the <em>maximal</em> block length possible, but the actual block lengths per segment will be smaller (proportional to the fraction of basepairs covered by that state in the genome segmentation).</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">gr_prime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootRanges.html">bootRanges</a></span><span class="op">(</span><span class="va">gr</span>, blockLength <span class="op">=</span> <span class="fl">50</span>, seg <span class="op">=</span> <span class="va">seg</span>,
                       proportionLength <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">plotGRanges</span><span class="op">(</span><span class="va">gr_prime</span><span class="op">)</span></code></pre></div>
<p><img src="segmented_boot_ranges_files/figure-html/toy-prop-1.png"><!-- --></p>
<p>Note that some ranges from adjacent states are allowed to be placed within different states in the bootstrap sample. This is because, during the random sampling of blocks of original data, a block is allowed to extend beyond the segmentation region of the state being sampled, and features from adjacent states are not excluded from the sampled block.</p>
</div>
</div>
<div class="section level1">
<h1 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h1>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.1 (2021-08-10)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.3 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] plotgardener_1.0.6          microbenchmark_1.4.9       </span>
<span class="co">##  [3] nullrangesData_1.0.0        InteractionSet_1.22.0      </span>
<span class="co">##  [5] SummarizedExperiment_1.24.0 MatrixGenerics_1.6.0       </span>
<span class="co">##  [7] matrixStats_0.61.0          ExperimentHub_2.2.0        </span>
<span class="co">##  [9] patchwork_1.1.1             plyranges_1.14.0           </span>
<span class="co">## [11] nullranges_1.1.4            excluderanges_0.99.6       </span>
<span class="co">## [13] AnnotationHub_3.2.0         BiocFileCache_2.2.0        </span>
<span class="co">## [15] dbplyr_2.1.1                EnsDb.Hsapiens.v86_2.99.0  </span>
<span class="co">## [17] ensembldb_2.18.2            AnnotationFilter_1.18.0    </span>
<span class="co">## [19] GenomicFeatures_1.46.1      AnnotationDbi_1.56.2       </span>
<span class="co">## [21] Biobase_2.54.0              GenomicRanges_1.46.1       </span>
<span class="co">## [23] GenomeInfoDb_1.30.0         IRanges_2.28.0             </span>
<span class="co">## [25] S4Vectors_0.32.3            BiocGenerics_0.40.0        </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.3             plyr_1.8.6                   </span>
<span class="co">##   [3] RcppHMM_1.2.2                 lazyeval_0.2.2               </span>
<span class="co">##   [5] BiocParallel_1.28.3           ggplot2_3.3.5                </span>
<span class="co">##   [7] digest_0.6.29                 yulab.utils_0.0.4            </span>
<span class="co">##   [9] htmltools_0.5.2               fansi_0.5.0                  </span>
<span class="co">##  [11] magrittr_2.0.1                memoise_2.0.1                </span>
<span class="co">##  [13] ks_1.13.2                     Biostrings_2.62.0            </span>
<span class="co">##  [15] pkgdown_2.0.1                 prettyunits_1.1.1            </span>
<span class="co">##  [17] colorspace_2.0-2              blob_1.2.2                   </span>
<span class="co">##  [19] rappdirs_0.3.3                textshaping_0.3.6            </span>
<span class="co">##  [21] xfun_0.28                     dplyr_1.0.7                  </span>
<span class="co">##  [23] crayon_1.4.2                  RCurl_1.98-1.5               </span>
<span class="co">##  [25] jsonlite_1.7.2                glue_1.5.1                   </span>
<span class="co">##  [27] gtable_0.3.0                  zlibbioc_1.40.0              </span>
<span class="co">##  [29] XVector_0.34.0                strawr_0.0.9                 </span>
<span class="co">##  [31] DelayedArray_0.20.0           scales_1.1.1                 </span>
<span class="co">##  [33] mvtnorm_1.1-3                 DBI_1.1.1                    </span>
<span class="co">##  [35] Rcpp_1.0.7                    xtable_1.8-4                 </span>
<span class="co">##  [37] progress_1.2.2                gridGraphics_0.5-1           </span>
<span class="co">##  [39] bit_4.0.4                     mclust_5.4.8                 </span>
<span class="co">##  [41] httr_1.4.2                    RColorBrewer_1.1-2           </span>
<span class="co">##  [43] speedglm_0.3-3                ellipsis_0.3.2               </span>
<span class="co">##  [45] pkgconfig_2.0.3               XML_3.99-0.8                 </span>
<span class="co">##  [47] farver_2.1.0                  sass_0.4.0                   </span>
<span class="co">##  [49] utf8_1.2.2                    DNAcopy_1.68.0               </span>
<span class="co">##  [51] ggplotify_0.1.0               tidyselect_1.1.1             </span>
<span class="co">##  [53] labeling_0.4.2                rlang_0.4.12                 </span>
<span class="co">##  [55] later_1.3.0                   munsell_0.5.0                </span>
<span class="co">##  [57] BiocVersion_3.14.0            tools_4.1.1                  </span>
<span class="co">##  [59] cachem_1.0.6                  generics_0.1.1               </span>
<span class="co">##  [61] RSQLite_2.2.9                 ggridges_0.5.3               </span>
<span class="co">##  [63] evaluate_0.14                 stringr_1.4.0                </span>
<span class="co">##  [65] fastmap_1.1.0                 yaml_2.2.1                   </span>
<span class="co">##  [67] ragg_1.2.1                    knitr_1.36                   </span>
<span class="co">##  [69] bit64_4.0.5                   fs_1.5.2                     </span>
<span class="co">##  [71] purrr_0.3.4                   KEGGREST_1.34.0              </span>
<span class="co">##  [73] mime_0.12                     pracma_2.3.6                 </span>
<span class="co">##  [75] xml2_1.3.3                    biomaRt_2.50.1               </span>
<span class="co">##  [77] compiler_4.1.1                filelock_1.0.2               </span>
<span class="co">##  [79] curl_4.3.2                    png_0.1-7                    </span>
<span class="co">##  [81] interactiveDisplayBase_1.32.0 tibble_3.1.6                 </span>
<span class="co">##  [83] bslib_0.3.1                   stringi_1.7.6                </span>
<span class="co">##  [85] highr_0.9                     desc_1.4.0                   </span>
<span class="co">##  [87] lattice_0.20-45               ProtGenerics_1.26.0          </span>
<span class="co">##  [89] Matrix_1.4-0                  vctrs_0.3.8                  </span>
<span class="co">##  [91] pillar_1.6.4                  lifecycle_1.0.1              </span>
<span class="co">##  [93] BiocManager_1.30.16           jquerylib_0.1.4              </span>
<span class="co">##  [95] data.table_1.14.2             bitops_1.0-7                 </span>
<span class="co">##  [97] httpuv_1.6.3                  rtracklayer_1.54.0           </span>
<span class="co">##  [99] R6_2.5.1                      BiocIO_1.4.0                 </span>
<span class="co">## [101] promises_1.2.0.1              KernSmooth_2.23-20           </span>
<span class="co">## [103] MASS_7.3-54                   assertthat_0.2.1             </span>
<span class="co">## [105] rprojroot_2.0.2               rjson_0.2.20                 </span>
<span class="co">## [107] withr_2.4.3                   GenomicAlignments_1.30.0     </span>
<span class="co">## [109] Rsamtools_2.10.0              GenomeInfoDbData_1.2.7       </span>
<span class="co">## [111] parallel_4.1.1                hms_1.1.1                    </span>
<span class="co">## [113] grid_4.1.1                    rmarkdown_2.11               </span>
<span class="co">## [115] shiny_1.7.1                   restfulr_0.0.13</span></code></pre>
</div>
<div class="section level1 unnumbered">
<h1 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-encode_exclude" class="csl-entry">
Amemiya, Haley M, Anshul Kundaje, and Alan P Boyle. 2019. <span>“The <span>ENCODE</span> Blacklist: Identification of Problematic Regions of the Genome.”</span> <em>Scientific Reports</em> 9 (1): 9354. <a href="https://doi.org/10.1038/s41598-019-45839-z" class="external-link">https://doi.org/10.1038/s41598-019-45839-z</a>.
</div>
<div id="ref-bickel_2010" class="csl-entry">
Bickel, Peter J., Nathan Boley, James B. Brown, Haiyan Huang, and Nancy R. Zhang. 2010. <span>“Subsampling Methods for Genomic Inference.”</span> <em>The Annals of Applied Statistics</em> 4 (4): 1660–97. <a href="https://doi.org/10.1214/10-%7BAOAS363%7D" class="external-link">https://doi.org/10.1214/10-{AOAS363}</a>.
</div>
<div id="ref-encode" class="csl-entry">
ENCODE. 2012. <span>“<span class="nocase">An integrated encyclopedia of DNA elements in the human genome</span>.”</span> <em>Nature</em> 489 (7414): 57–74. <a href="https://doi.org/10.1038/nature11247" class="external-link">https://doi.org/10.1038/nature11247</a>.
</div>
<div id="ref-ensembl2021" class="csl-entry">
Howe, Kevin L, Premanand Achuthan, James Allen, Jamie Allen, Jorge Alvarez-Jarreta, M Ridwan Amode, Irina M Armean, et al. 2020. <span>“<span>Ensembl 2021</span>.”</span> <em>Nucleic Acids Research</em> 49 (D1): D884–91. <a href="https://doi.org/10.1093/nar/gkaa942" class="external-link">https://doi.org/10.1093/nar/gkaa942</a>.
</div>
<div id="ref-Kramer2021" class="csl-entry">
Kramer, Nicole E, Eric S Davis, Craig D Wenger, Erika M Deoudes, Sarah M Parker, Michael I Love, and Douglas H Phanstiel. 2021. <span>“<span class="nocase">Plotgardener: Cultivating precise multi-panel figures in R</span>.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.09.08.459338" class="external-link">https://doi.org/10.1101/2021.09.08.459338</a>.
</div>
<div id="ref-Lee2019" class="csl-entry">
Lee, Stuart, Dianne Cook, and Michael Lawrence. 2019. <span>“Plyranges: A Grammar of Genomic Data Transformation.”</span> <em>Genome Biology</em> 20 (1): 4. <a href="https://doi.org/10.1186/s13059-018-1597-8" class="external-link">https://doi.org/10.1186/s13059-018-1597-8</a>.
</div>
<div id="ref-dnacopy" class="csl-entry">
Olshen, A. B., E. S. Venkatraman, R. Lucito, and M. Wigler. 2004. <span>“<span class="nocase"><span>C</span>ircular binary segmentation for the analysis of array-based <span>D</span><span>N</span><span>A</span> copy number data</span>.”</span> <em>Biostatistics</em> 5 (4): 557–72.
</div>
<div id="ref-ensembldb" class="csl-entry">
Rainer, Johannes, Laurent Gatto, and Christian X Weichenberger. 2019. <span>“<span class="nocase">ensembldb: an R package to create and use Ensembl-based annotation resources</span>.”</span> <em>Bioinformatics</em> 35 (17): 3151–53. <a href="https://doi.org/10.1093/bioinformatics/btz031" class="external-link">https://doi.org/10.1093/bioinformatics/btz031</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://twitter.com/wancenm" class="external-link">Wancen Mu</a>, <a href="https://ericscottdavis.com" class="external-link">Eric Davis</a>, <a href="http://phanstiel-lab.med.unc.edu" class="external-link">Douglas Phanstiel</a>, <a href="https://stuartlee.org" class="external-link">Stuart Lee</a>, <a href="https://chanzuckerberg.com" class="external-link">CZI</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
